---
title: 'Genomics: DNA&#8217;s master craftsmen'
author: 悟道
excerpt: |
  Published online 3 November 2010 | Nature 468, 22-25 (2010) | doi:10.1038/468022a
  
  News Feature
  Genomics: DNA's master craftsmen
  
  Behind the walls of the J. Craig Venter Institute, Ham Smith and Clyde Hutchison quietly worked to bring a synthetic cell to life.
layout: post
permalink: /?p=325
categories:
  - bio
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

Sitting in adjoining offices on the second floor of the J. Craig Venter Institute (JCVI) in San Diego, California, Ham Smith and Clyde Hutchison carry on a fragmented conversation through their open doors.

&#8220;Clyde, did you do your timesheet?&#8221; says Smith. &#8220;It&#8217;s due in 12 minutes.&#8221;

Hutchison pauses: &#8220;No, it was due three hours ago.&#8221;

They work at their computers for a minute in silence. Both are dressed in green short-sleeved shirts, tan trousers and black shoes, although they swear the wardrobe coordination is an accident. &#8220;The market finished flat,&#8221; says Smith.

The dialogue is almost constant between the two men — friends and collaborators who are rarely seen apart. They are the veteran DNA craftsmen at the JCVI, a non-profit research organization founded by the provocative genome scientist who gave it his name. Originally from academia, the pair has taken advantage of Venter&#8217;s resources and unparalleled salesmanship to pursue ambitious projects in synthetic biology, research that Venter says could enable faster vaccine production and the development of organisms that churn out precursors to fuel.

In May, the JCVI team announced that it had built a 1-million-base-pair genome — the longest working piece of chemically synthesized DNA yet assembled — and used it to restart a bacterial cell<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B1">1</a></sup>. Although some scientists disagree on whether the resulting microorganism, called &#8216;Synthia&#8217; in the popular press, is indeed &#8216;synthetic&#8217; — the synthesized genome sequence was cribbed from a related bacterial species rather than being built to a novel design — few deny the technical skill demonstrated by such work. &#8220;The ability to synthesize and put together so many nucleotides without a mistake really requires guys on the level of Smith and Hutchison,&#8221; says David Botstein, a geneticist at Princeton University in New Jersey who has worked with Smith. &#8220;I don&#8217;t think many other people could have done it.&#8221;

The two are acclaimed for their pioneering work. Smith shared a Nobel Prize in Physiology or Medicine in 1978 for his discovery of a restriction enzyme<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B2">2</a></sup>— a protein that cuts DNA at specific sites. In the 1970s, Hutchison helped to determine the first full sequence of a DNA molecule and co-developed site-directed mutagenesis<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B3">3</a></sup>, a technique that enables researchers to make targeted changes to DNA sequences. The methods Smith and Hutchison helped to develop underpin much of the work done today in molecular genetics. &#8220;They&#8217;re both regarded as scientist&#8217;s scientists,&#8221; says David Baltimore, a molecular biologist at the California Institute of Technology in Pasadena. &#8220;They&#8217;ve both done enormously important work in basic science.&#8221;

Now 79 and 71, respectively, Smith and Hutchison have become intellectual partners and close friends. Smith, who looms over Hutchison, is slouchy, often sporting a half smile somewhere between amusement and embarrassment; Hutchison is more deliberate, precise in movement and dry in delivery. They are still pushing the possibilities of their field and, stints at the desk aside, they try to spend about half their working hours at the bench — something that helps them maintain &#8220;a real perspective on what is possible&#8221;, says Hutchison. This comes in handy when tasked with meeting the sometimes audacious goals set by their boss. &#8220;At one level, each of them operates like a septuagenarian postdoc,&#8221; says John Glass, a synthetic biologist at the JCVI&#8217;s other campus in Rockville, Maryland.

## Three&#8217;s company

When Smith and Venter first met in 1993, Venter already had a reputation. He was still five years away from challenging the public effort to sequence the human genome, but the US National Institutes of Health, where Venter worked until 1992, had filed patent applications on DNA fragments sequenced by Venter&#8217;s team. The move didn&#8217;t go down well with other scientists, nor did his tendency to antagonize scientific rivals. &#8220;He was commonly called an asshole,&#8221; says Smith. But over drinks in a bar in Spain, the two discovered common ground: both started in medicine and had served in the navy. Venter invited Smith, then a professor at Johns Hopkins University School of Medicine in Baltimore, Maryland, out for dinner with friends, and the group got drunk. &#8220;Almost the first instant that I actually met him, I liked him,&#8221; says Smith.

Smith joined the scientific advisory council of the non-profit organization The Institute for Genomic Research (TIGR) in Gaithersburg, Maryland, which Venter had founded, and began collaborating scientifically with Venter, despite concerns of academic colleagues that the association might taint Smith&#8217;s career. Their teams worked on determining the first genome sequence of an independently living organism, the bacterium Haemophilus influenzae<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B4">4</a></sup> — in which Smith and co-workers had originally discovered the HindII restriction enzyme. They chopped the 1.8-million-base-pair genome into pieces, sequenced the fragments and then assembled them computationally into a genome sequence. It was the first time the technique, called shotgun sequencing, had been used on such a large DNA molecule. Smith&#8217;s excitement, colleagues say, outweighed practical considerations. Jean-François Tomb, a former research associate at Johns Hopkins, recalls members of Smith&#8217;s lab worrying about a lack of funding for the project, but Smith was only interested in the science. &#8220;He said, &#8216;Look, you sequence the genome once and it&#8217;s forever,&#8217;&#8221; says Tomb.

<div id="p3">
  <blockquote>
    <div>
      “All three of them are more about just doing the home-run experiment.”
    </div>
  </blockquote>
  
  <p>
    <cite id="n1"><br /> </cite>
  </p>
</div>

When the H. influenzae genome was nearly done, Venter wanted to sequence another right away. Smith said that they should try something small, and thought of Hutchison, then at the University of North Carolina at Chapel Hill. Hutchison was studying Mycoplasma genitalium — a bacterium thought to have the smallest genome, at half a million base pairs, of any free-living organism. Venter liked the idea. &#8220;He said, well, as soon as we leave lunch, why don&#8217;t you call him up?&#8221; says Smith.

Hutchison agreed to help and sent Smith 10 micrograms of M. genitalium DNA. They finished the sequence in about two months<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B5">5</a></sup>. &#8220;I was very pleased,&#8221; says Hutchison. New techniques appealed to him, and he often pushed his team to develop better methods. &#8220;From the beginning, Clyde was into high-throughput,&#8221; says Mike Conrad, a former postdoc in Hutchison&#8217;s University of North Carolina lab. &#8220;He liked stuff that was fast.&#8221;

When Hutchison took a sabbatical at TIGR in 1996, he, Smith and Venter began to discuss the idea of developing a cell with the minimum genome needed to survive. Hutchison was already investigating which genes M. genitalium could live without<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B6">6</a></sup>, but he knew that deleting multiple genes simultaneously from this bacterium was technically difficult. The threesome speculated that they might need to synthesize whole candidate genomes and test them in recipient cells.

Hutchison again collaborated with Smith in 2003 at Venter&#8217;s latest non-profit institute, the Institute for Biological Energy Alternatives (IBEA) in Rockville. Their team synthesized the 5,000-base-pair genome of the bacteriophage ΦX174 (ref. <sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B7">7</a></sup>). Hutchison had helped determine its sequence in the 1970s, and the small size made it convenient for trying out synthetic techniques. Smith and Hutchison had very different experimental styles, recalls team member Cindi Pfannkoch. &#8220;Clyde likes to plan everything,&#8221; she says, whereas Smith practices more casual &#8216;bathtub biochemistry&#8217;. In spite of this, the two got along. &#8220;They speak the same language,&#8221; she says.

## Taming the cell

In 2005, Hutchison started full time at the JCVI, which was formed by a merger of the IBEA and other Venter organizations. Leaving university life has disadvantages: &#8220;I can&#8217;t do whatever I want,&#8221; says Hutchison. &#8220;We&#8217;re not totally independent agents.&#8221; But the trio&#8217;s interest in big scientific challenges has kept them together. &#8220;I think all three of them are more about just doing the home-run experiment,&#8221; says Dan Gibson, a synthetic biologist at the JCVI in Rockville.

<div id="p1">
  <blockquote>
    <div>
      “We didn&#8217;t know it would be possible till it was done.”
    </div>
  </blockquote>
  
  <p>
    <cite id="n2"><br /> </cite>
  </p>
</div>

The synthetic-cell project picked up steam in 2005 as more researchers joined Venter&#8217;s synthetic-biology team, which eventually comprised some two dozen scientists. The project evolved into a two-pronged effort: one group focused on constructing a synthetic M. genitalium genome, while the other tried to transplant natural genomes into cells of different Mycoplasma strains and species (see [&#8216;The path to a synthetic cell&#8217;][1]).

[ ][2]

<div>
  <a href="http://www.nature.com/news/2010/101103/full/468022a/box/1.html"><img src="http://www.nature.com/news/2010/101103/images/468022a-i5.0.jpg" alt="" />Click for a larger version.</a>
</div>

<a name="a"></a>

Smith and Hutchison, who worked on genome construction, were circumspect about the chance of success. &#8220;We didn&#8217;t know it would be possible till it was done,&#8221; says Hutchison. Technical challenges loomed. The large DNA segments might break; the slow growth rate of M. genitalium limited the pace of progress; and — most crucially — rebooting a cell with a new genome had never been done before.

Although often seen just as the public face of the JCVI, Venter contributed to the science. While Smith and Hutchison worked out the details, Venter made key strategic decisions. At first, the team tried assembling pieces of the M. genitalium genome from short DNA fragments rather than ordering longer, prefabricated, but more expensive segments from a DNA synthesis company. But progress was slow. &#8220;After we&#8217;d been working on this a couple of months, Craig comes into the lab and says, &#8216;Ham, how many pieces do you have put together so far?&#8217;&#8221; says Smith. &#8220;And I said, &#8216;Well, we haven&#8217;t got any of them yet.&#8217; And he says, &#8216;All right, we&#8217;re going to order them&#8217;.&#8221;

The genome-transplantation group was having no luck either. Carole Lartigue, a postdoc on the team, worked on the problem for two years &#8220;with nothing but failure&#8221;, says Glass, who oversaw the research. Lartigue finally got the first evidence of successful transplantation in late 2006, and the following year the team announced that it had managed to transplant a natural — not synthetic — genome from Mycoplasma mycoides into cells of the related species Mycoplasma capricolum, changing their identity<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B8">8</a></sup>. Although the paper was announced to great fanfare, the mood at the JCVI wasn&#8217;t always so jubilant. A few months after publication, Smith came into the lab distraught. He feared that some M. mycoides cells, from which the donor genome was originally isolated, might have become patched up and revived when mixed with the recipient cells — an idea he called the &#8216;punctured tyre&#8217; hypothesis. Resurrected cells could have been mistaken for transplants.

Hutchison, who is known for being meticulous about experimental controls, says he &#8220;thought the evidence was pretty good that wasn&#8217;t the case&#8221;. But Smith&#8217;s doubts were not laid to rest until late in 2008, when the team started up M. capricolum cells with a natural M. mycoides genome that had first been inserted into and modified in yeast — a process that &#8216;purified&#8217; the genome of any cellular remnants of its original host<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B9">9</a></sup>. &#8220;That was absolute proof,&#8221; says Smith.

By now, Smith and Hutchison were working from the JCVI&#8217;s new lab in San Diego — a move prompted, according to Smith, by Mary-land&#8217;s freezing winters. In 2003, Smith had slipped on the ice and broken his leg. A couple of years later, while he and Venter were walking through the sleet in Rockville, &#8220;I turned to Craig and said, &#8216;Why do we live here?&#8217;&#8221; says Smith. &#8220;He said, &#8216;That&#8217;s a good point.&#8217;&#8221; Venter set up a lab near the University of California, San Diego, his alma mater, and Smith and Hutchison moved west. &#8220;I came out because Ham was coming out,&#8221; says Hutchison.

By this time, JCVI researchers had also chemically synthesized the M. genitalium genome. Starting with segments of 5,000 to 7,000 base pairs made by DNA synthesis companies, the researchers connected them into progressively larger pieces of DNA, first in vitro, and then in yeast<sup><a href="http://www.nature.com/news/2010/101103/full/468022a.html#B10">10</a></sup>. But they still couldn&#8217;t successfully transplant the M. genitalium genome, and the organism&#8217;s slow growth rate meant that they had to wait at least a month to see the results of an experiment. &#8220;We just didn&#8217;t know if it would ever work,&#8221; says molecular biologist Gwynedd Benders, a former team member. &#8220;It&#8217;s like, are we just going to keep hammering at this?&#8221;

As early as 2007, Venter had suggested switching the donor species: synthesizing the M. mycoides genome — which was roughly twice as large as M. genitalium &#8216;s — and transplanting it into M. capricolum. Mycoplasma mycoides grew faster, and the team had already managed to transplant the bacterium&#8217;s natural genome. The researchers had been reluctant to switch because they didn&#8217;t yet know whether the M. mycoides genome could be transplanted from yeast, where the synthesized genome would need to be assembled, into bacteria. But once they had successfully done this with the natural genome, Venter&#8217;s idea seemed more attractive. In late 2008, Venter discussed a strategy for the M. mycoides genome synthesis with Smith, and Smith e-mailed Gibson, &#8220;Craig wants this done PDQ&#8221; — pretty damn quick. &#8220;It&#8217;s that kind of thing where Craig is absolutely essential,&#8221; says Smith. &#8220;He was pushing it. I would have dragged for months.&#8221;

## Waiting game

But synthesizing the M. mycoides genome presented new problems. Pieces of this genome above a certain size didn&#8217;t replicate well in Escherichia coli, the bacterium used to amplify the DNA. Even determining the M. mycoides genome sequence proved difficult because of repetitive DNA sequences. &#8220;It just seemed to go on for months and months,&#8221; says Benders.

After the M. mycoides genome was finally assembled, the team endured some suspense-filled weekends. Transplantation experiments were performed on Fridays, and Gibson checked for the blue bacterial colonies that would indicate success on Monday mornings. Week after week, no blue colonies appeared, and finally a single mutation was discovered in one of the synthetic DNA segments. This was corrected and transplantation tried again on Friday 26 March 2010. &#8220;That was a really, really long weekend,&#8221; says Gibson.

On the Monday morning around 6 a.m. Gibson found a single blue colony and e-mailed Venter, Smith, Hutchison and Glass. Knowing tests were still needed to confirm the result, Gibson says he was &#8220;sweating&#8221; all day about the possibility of contamination, even as they celebrated with champagne. But Smith, who had told him to wake Venter that morning with the news, was more sanguine. Gibson says, &#8220;He just knew it was real&#8221;.

Since the announcement, the team has fielded criticism for calling the resulting cell &#8216;synthetic&#8217; when the genome was essentially a replica of a natural genome and required an existing recipient cell. Hutchison argues that &#8216;synthetic&#8217; simply means &#8216;chemically synthesized&#8217;, not newly designed, and recipient cell contents are eventually replaced. &#8220;You&#8217;d like to design a genome from scratch,&#8221; he says. &#8220;You&#8217;d like to put it into a cytoplasm that you built up from scratch. But we&#8217;re trying to do something we can do.&#8221; Although many synthetic-biology researchers are tweaking existing genetic elements, assembling them into new combinations, and inserting the &#8216;circuits&#8217; into different organisms, few aspire to design entirely new genes.

On a July morning, Smith and Hutchison sit together in an auditorium at the JCVI&#8217;s San Diego building listening to a presentation by summer intern Nico Enriquez. Members of the synthetic-biology team, whose lab is down the corridor from Smith and Hutchison&#8217;s offices, are scattered in the audience; researchers in Rockville watch the talk by videoconference. The synthetic-biology team is now attempting to develop the &#8216;minimal cell&#8217;: Gibson and his colleagues are building new versions of the synthetic M. mycoides genome with genes removed, then transplanting the edited genomes into recipient cells and monitoring colony size and growth. Enriquez presents methods proposed by Smith to assess cells&#8217; growth rates by measuring DNA and protein levels, and initial results look promising. &#8220;Seems like Ham&#8217;s got something right again,&#8221; says Enriquez.

At the end of his talk, Enriquez shows a picture of a pair of frolicking otters. &#8220;They kind of remind me of Ham and Clyde,&#8221; he says to laughter in the audience. &#8220;The tall one is Ham, and the shorter one is Clyde — always together.&#8221;

<div id="p2">
  <blockquote>
    <div>
      “I just want to understand, thoroughly, one cell.”
    </div>
  </blockquote>
  
  <p>
    <cite id="n3"><br /> </cite>
  </p>
</div>

Once the researchers have a minimal genome, they hope to determine the function of every uncharacterized gene and build a computer model that predicts the cell&#8217;s responses to genetic changes. But such a system isn&#8217;t necessarily more informative than an organism with a larger genome, argues geneticist George Church at Harvard Medical School in Boston, Massachusetts. Although the cell will yield some scientific insights, he says, they will probably be specific to Mycoplasma. &#8220;As you delete these things, you&#8217;ll end up with a cell that is weaker and weaker, less and less industrially useful, and less and less relevant to sophisticated organisms,&#8221; he says.

Smith and Hutchison agree that Mycoplasma is unlikely to be used industrially as it is expensive to grow, but say that lessons learned from developing and studying this system might apply to organisms better suited for commercial purposes. Hutchison says a smaller system will be easier to understand, and he predicts that some gene deletions might actually make the cell grow faster. Smith notes that scientists initially couldn&#8217;t see all the uses of restriction enzymes, which have proved essential for manipulating DNA. And in any case, he is largely driven by curiosity. &#8220;I just want to understand, thoroughly, one cell,&#8221; he says. Venter, too, emphasizes that the minimal cell is meant to be a proof of concept. &#8220;It&#8217;s truly basic science,&#8221; he says.

JCVI researchers are investigating whether their genome-transplantation techniques can be extended to other, more complex, bacteria, such as cyanobacteria. The team is also trying to make large-scale changes to cyanobacterial genomes, with the eventual goal of enabling production of valuable chemicals, says Glass. On the more commercial side, Venter and Smith&#8217;s company, Synthetic Genomics, based in La Jolla, California, has filed patent applications on the JCVI&#8217;s methods, and aims to engineer algae that produce hydrocarbons that can be converted into fuel. A separate company, Synthetic Genomics Vaccines, set up by the JCVI and Synthetic Genomics, is working with pharmaceutical company Novartis, based in Basel, Switzerland, on ways of making flu vaccine production more efficient.

Although other synthetic biologists are already using the JCVI&#8217;s techniques to assemble pieces of DNA, many agree that it will be a while before anyone designs a whole genome from scratch. Right now, the quickest route to industrial application is unquestionably the modification of existing organisms, says Venter. But, he says, &#8220;the future will be designing and making whole new species&#8221;.

<div>
  <p>
    ADVERTISEMENT
  </p>
  
  <noscript>
    <a target="_blank" href="http://ad.doubleclick.net/click%3Bh%3Dv8/3a48/3/0/%2a/g%3B229338440%3B0-0%3B0%3B9916591%3B4307-300/250%3B38518556/38536313/1%3B%3B%7Esscs%3D%3fhttp://www.nature.com/scitable/topics"><img src=""  border="0" alt="" /></a>
  </noscript>
  
  <noscript>
    <div>
      <a href="http://ad.doubleclick.net/jump/news@nature.com/;abr=!NN2;artid=article-one;pos=left;sz=300x250;ptile=2;ord=123456789?"><img src="http://ad.doubleclick.net/ad/news@nature.com/;abr=!NN2;tile=1;ord=123456789?" alt="Advertisement" /></a>
    </div>
  </noscript>
</div>

It is now down to Smith and Hutchison to make Venter&#8217;s ideas a reality — their working styles distinct, but complementary. At a lab meeting, Smith sketches a plan on a whiteboard for an experiment to detect very small changes in bacterial growth rate, which will be necessary to compare M. mycoides strains with different genome versions.

&#8220;You should do multiple dilutions each time,&#8221; says Hutchison.

Smith hesitates. &#8220;I want it to be very simple, so …&#8221;

&#8220;Yeah, but maybe in working out how to do it, you need to do something that&#8217;s not quite as simple,&#8221; says Hutchison.

The two have no plans to retire yet. Most weekdays they take a walk together in the hills around their office, which overlook Interstate Highway 5, ruminating on ideas and keeping an eye out for rattlesnakes. &#8220;I think this is the pinnacle of my career,&#8221; says Smith, who wants to keep working for at least another four or five years. Hutchison adds, &#8220;But maybe we&#8217;ll do something else next.&#8221;

Roberta Kwok is a freelance writer in the San Francisco Bay Area.

<ul id="article-refrences">
  <li>
    <h2>
      References
    </h2>
    
    <ol>
      <li id="B1">
        <a name="B1"></a>Gibson, D. G. et al. Science 329, 52-56 (2010).
      </li>
      <li id="B2">
        <a name="B2"></a>Smith, H. O. & Wilcox, K. W. J. Mol. Biol. 51, 379-391 (1970).
      </li>
      <li id="B3">
        <a name="B3"></a>Hutchison, C. A. et al. J. Biol. Chem. 253, 6551-6560 (1978).
      </li>
      <li id="B4">
        <a name="B4"></a>Fleischmann, R. D. et al. Science 269, 496-512 (1995).
      </li>
      <li id="B5">
        <a name="B5"></a>Fraser, C. M. et al. Science 270, 397-403 (1995).
      </li>
      <li id="B6">
        <a name="B6"></a>Hutchison, C. A. et al. Science 286, 2165-2169 (1999).
      </li>
      <li id="B7">
        <a name="B7"></a>Smith, H. O. , Hutchison, C. A. , Pfannkoch, C. & Venter, J. C. Proc. Natl Acad. Sci. USA 100, 15440-15445 (2003).
      </li>
      <li id="B8">
        <a name="B8"></a>Lartigue, C. et al. Science 317, 632-638 (2007).
      </li>
      <li id="B9">
        <a name="B9"></a>Lartigue, C. et al. Science 325, 1693-1696 (2009).
      </li>
      <li id="B10">
        <a name="B10"></a>Gibson, D. G. et al. Science 319, 1215-1220 (2008).
      </li>
    </ol>
  </li>
</ul>

 [1]: http://www.nature.com/news/2010/101103/full/468022a.html#a
 [2]: http://www.nature.com/news/2010/101103/full/468022a/box/1.html