---
title: Alternative splicing
author: 悟道
layout: post
permalink: /?p=2479
categories:
  - literature
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

# Status:

It is estimated that greater than 90% of human genes undergo alternative splicing of pre-mRNA<sup><a id="ref-link-1" title="Wang, E. T. et al. Alternative isoform regulation in human tissue transcriptomes. Nature 456, 470-476 (2008)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref1">1</a>, <a id="ref-link-2" title="Pan, Q., Shai, O., Lee, L. J., Frey, B. J. & Blencowe, B. J. Deep surveying of alternative splicing complexity in the human transcriptome by high-throughput sequencing. Nature Genet. 40, 1413-1415 (2008)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref2">2</a></sup> and aberrant splicing has been implicated in a number of human diseases<sup><a id="ref-link-3" title="Tazi, J., Bakkour, N. & Stamm, S. Alternative splicing and disease. Biochim. Biophys. Acta 1792, 14-26 (2009)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref3">3</a></sup>.

# Effector:

1.Alternative splicing decisions are determined by the ability of weak splice sites to effectively compete with strong splice sites for detection by the spliceosome<sup><a id="ref-link-4" title="Matlin, A. J., Clark, F. & Smith, C. W. Understanding alternative splicing: towards a cellular code. Nature Rev. Mol. Cell Biol. 6, 386-398 (2005)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref4">4</a></sup>. The balance between splice site selection is principally influenced by two variables<sup><a id="ref-link-5" title="Han, J., Xiong, J., Wang, D. & Fu, X. D. Pre-mRNA splicing: where and when in the nucleus. Trends Cell Biol. 21, 336-343 (2011)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref5">5</a></sup>: (1) the availability of splicing factors that detect enhancer or silencer sequences encoded within nascent RNA<sup><a id="ref-link-6" title="Matlin, A. J., Clark, F. & Smith, C. W. Understanding alternative splicing: towards a cellular code. Nature Rev. Mol. Cell Biol. 6, 386-398 (2005)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref4">4</a>, <a id="ref-link-7" title="Singh, R. & Valcarcel, J. Building specificity with nonspecific RNA-binding proteins. Nature Struct. Mol. Biol. 12, 645-653 (2005)" href="http://www.nature.com/nature/journal/v479/n7371/full/nature10442.html#ref6">6</a></sup> and (2) the rate of RNA polymerase II (pol II) transcription elongation, wherein a slow rate favours co-transcriptional spliceosome assembly at weak splice sites

2.