---
title: 20120905-
author: 悟道
layout: post
permalink: /?p=2475
categories:
  - literature
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

### [DNA methylation at the exon 2 loci is associated with reduced RNApII transcriptional elongation][1]

<p id="p-45">
  Having demonstrated the association of DNA methylation with <em>PolgA</em> expression and mtDNA copy number<em>,</em> we further explored this relationship by examining transcription factor binding sites within exon 2 of <em>PolgA</em>. Using the JASPAR CORE transcription factor database (<a href="http://jaspar.genereg.net/cgi-bin/jaspar_db.pl">http://jaspar.genereg.net/cgi-bin/jaspar_db.pl</a>) (<a id="xref-ref-47-1" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-47">47–50</a>), we identified an evolutionary conserved putative CTCF binding site within this region (<a href="http://nar.oxfordjournals.org/cgi/content/full/gks770/DC1">Supplementary Figure S5A</a>). Since CTCF regulates transcription (<a id="xref-ref-41-2" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-41">41</a>,<a id="xref-ref-51-1" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-51">51</a>) and DNA methylation inhibits its binding (<a id="xref-ref-41-3" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-41">41</a>), we determined whether CTCF binding regulates <em>PolgA</em> expression levels. Using the ChIP assay, we examined the enrichment of CTCF in CC9<em><sup>mus</sup></em> ESCs and MEFs cells at the exon 2 loci (<a href="http://nar.oxfordjournals.org/cgi/content/full/gks770/DC1">Supplementary Figure S5B</a>). Both cell types showed similar levels of CTCF enrichment within exon 2 and at sites upstream and downstream (<a href="http://nar.oxfordjournals.org/cgi/content/full/gks770/DC1">Supplementary Figure S5C</a>). Since CC9<em><sup>mus</sup></em> and MEFs cells display different DNA methylation profiles and <em>PolgA</em> expression signatures, these observations suggest that CTCF is not involved in regulating <em>PolgA</em> expression.
</p>

<p id="p-46">
  DNA methylation within coding sequences has been shown to alter transcript levels by influencing the recruitment or the transcriptional activation of RNApII (<a id="xref-ref-25-2" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-25">25</a>,<a id="xref-ref-26-1" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-26">26</a>,<a id="xref-ref-41-4" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#ref-41">41</a>). In order to determine how DNA methylation within exon 2 of <em>PolgA</em> modulates RNApII function, we quantified the enrichment of RNApII and the elongating RNApII phosphorylated at serine 2 of the carboxy-terminal domain (RNApIIS2). RNApII occupancy in CC9<em><sup>mus</sup></em>, MEF, NSC-CC9<em><sup>mus</sup></em> and heart samples was analysed upstream of the transcription start site within exon 2, and two sites downstream (<a id="xref-fig-6-1" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#F6">Figure 6</a>A). In CC9<em><sup>mus</sup></em> cells, RNApII (<em>P</em> < 0.001) and RNApIIS2 (<em>P</em> < 0.01) enrichment was significantly increased downstream of exon 2 (<a id="xref-fig-6-2" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#F6">Figure 6</a>B), while in MEF cells no differences were observed upstream or downstream of exon 2 (<a id="xref-fig-6-3" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#F6">Figure 6</a>C). Furthermore, in NSC-differentiated CC9<em><sup>mus</sup></em> cells, there was an initial increase in RNApIIS2 enrichment at a proximal site downstream of exon 2 (<em>P</em> < 0.05), although this returned to similar levels observed at exon 2 further downstream (<a id="xref-fig-6-4" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#F6">Figure 6</a>D). In heart tissue, there was no difference in enrichment for RNApII between exon 2 and downstream sites, whilst RNApIIS2 enrichment was decreased downstream of exon 2 (<a id="xref-fig-6-5" href="http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full#F6">Figure 6</a>E; <em>P</em> < 0.05). These results indicate that DNA methylation downstream of the <em>PolgA</em> transcription start site reduces transcription by RNApII at the elongation stage.
</p>

&nbsp;

# [Conserved role of intragenic DNA methylation in regulating alternative promoters][2]

Because many genes have alternative promoters, classically located upstream of the translation start site but also commonly present within genes<sup><a id="ref-link-37" title="Kimura, K. et al. Diversification of transcriptional modulation: large-scale identification and characterization of putative alternative promoters of human genes. Genome Res. 16, 55-65 (2006)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref15">15</a></sup>, we reasoned that a major function of the frequent, tissue-specific and conserved intragenic methylation may be to regulate the activity of such alternative promoters, as shown in two genes recently<sup><a id="ref-link-38" title="Rauch, T. A. et al. A human B cell methylome at 100-base pair resolution. Proc. Natl Acad. Sci. USA 106, 671-678 (2009)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref5">5</a>, <a id="ref-link-39" title="Irizarry, R. A. et al. The human colon cancer methylome shows similar hypo- and hypermethylation at conserved tissue-specific CpG island shores. Nature Genet. 41, 178-186 (2009)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref22">22</a></sup>. To address this hypothesis genome-wide, we determined whether the CGI loci overlap with sites of transcription initiation and/or with histone methylation marks typically found in association with 5′ promoters.

First, we assessed the relationship between the methylation status of CGIs in the human brain with CAGE tag data sets from several human tissues<sup><a id="ref-link-40" title="Carninci, P. et al. Genome-wide analysis of mammalian promoter architecture and evolution. Nature Genet. 38, 626-635 (2006)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref12">12</a>, <a id="ref-link-41" title="Valen, E. et al. Genome-wide detection and analysis of hippocampus core promoters using DeepCAGE. Genome Res. 19, 255-265 (2009)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref23">23</a></sup>. CAGE tags are derived from mRNA sequenced in the proximity of the 5′-cap site and those tags that map onto unique genomic regions correspond to potential transcriptional start sites<sup><a id="ref-link-42" title="The FANTOM Consortium. The transcriptional landscape of the mammalian genome. Science 309, 1559-1563 (2005)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref11">11</a>, <a id="ref-link-43" title="Carninci, P. et al. Genome-wide analysis of mammalian promoter architecture and evolution. Nature Genet. 38, 626-635 (2006)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref12">12</a>, <a id="ref-link-44" title="Kim, T. H. et al. A high-resolution map of active promoters in the human genome. Nature 436, 876-880 (2005)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref13">13</a>, <a id="ref-link-45" title="Kapranov, P. et al. Examples of the complex architecture of the human transcriptome revealed by RACE and high-density tiling arrays. Genome Res. 15, 987-997 (2005)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref14">14</a>, <a id="ref-link-46" title="Kimura, K. et al. Diversification of transcriptional modulation: large-scale identification and characterization of putative alternative promoters of human genes. Genome Res. 16, 55-65 (2006)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref15">15</a>, <a id="ref-link-47" title="Carninci, P. Tagging mammalian transcription complexity. Trends Genet. 22, 501-510 (2006)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref24">24</a></sup>, or in a few cases may be derived from post-transcriptionally processed RNAs<sup><a id="ref-link-48" title="Affymetrix/Cold Spring Harbor Laboratory ENCODE Transcriptome Project. Post-transcriptional processing generates a diversity of 5[prime]-modified long and short RNAs. Nature 457, 1028-1032 (2009)" href="http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html#ref25">25</a></sup>. The presence of CAGE tags from one or more tissue types suggests the underlying genomic sequence harbours a promoter, the activity of which depends on the cellular context and epigenetic status.

 [1]: http://nar.oxfordjournals.org/content/early/2012/08/31/nar.gks770.full
 [2]: http://www.nature.com/nature/journal/v466/n7303/full/nature09165.html