---
title: 文章写作中的方法部分
author: 悟道
layout: post
permalink: /?p=2376
categories:
  - literature
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

### Hierarchical Clustering {#tdbox1sec6.1}

<p id="p0310">
  To determine the overall tendencies of change in gene expression and the connection between different populations, we hierarchically clustered RNA-seq data of these 3,697 selected genes from all 11 samples (using normalized data, biological replicates and triplicates were treated independently, <a id="ancbfig1" href="http://www.sciencedirect.com/science/article/pii/S0092867412002930#fig1">Figure 1</a>B). Hierarchical clustering was performed along both dimensions with sample similarities clustered first, and then genes. Euclidean distance and complete linkage were used (MATLAB 7.10.0). Separately, two-dimensional hierarchical clustering was also performed on 379 differentially expressed transcription factors (<a id="ancbfig1" href="http://www.sciencedirect.com/science/article/pii/S0092867412002930#fig1">Figure 1</a>C). <em>【Dynamic Transformations of Genome-wide Epigenetic Marking and Transcriptional Control Establish T Cell Identity】</em>
</p>

### K-Means Clustering {#tdbox1sec6.2}

<p id="p0315">
  To profile and categorize the behavior of clusters of similarly regulated genes during early T cell development, we first normalized individual mRNA data for the 3,697 selected genes by the corresponding geometric mean of five stages, and then performed K-means clustering analysis on the results after log<sub>2</sub> transformation (<a id="ancbfigs5" href="http://www.sciencedirect.com/science/article/pii/S0092867412002930#figs5">Figure S5</a>). K was set at 25 and squared Euclidean distance was used (MATLAB 7.10.0). <em>【Dynamic Transformations of Genome-wide Epigenetic Marking and Transcriptional Control Establish T Cell Identity】</em>
</p>

&nbsp;