---
title: SNP
author: 悟道
layout: post
permalink: /?p=2967
categories:
  - seq
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

1.http://www.biostars.org/p/65812/  
My variation-toolkit contains a tool named &#8220;ucscgeneps&#8221;. It reads a tab-delimited file CHROM/POS/(SAMPLE), sends a mysql query to the UCSC database and displays the genes as a Postscript file:

<pre class="brush: bash; title: ; notranslate" title="">cat sample2vcf.tsv | tr -d ' ' |\
  scanvcf |\
  awk -F '      ' '($3==".")' |\
  normalizechrom |\
  sort -t '  ' -k1,1 -k2,2n -k11,11 |\
  head -n 10000 |\
  ucscgenesps  -s 11 &gt; result.ps
</pre>

2.LoFreq: a fast and sensitive variant-caller for inferring single-nucleotide variants from NGS data

http://www.biostars.org/p/54888/

3.http://www.biostars.org/p/65934/

4.Perhaps you should read my samtools paper, where I did an experiment on finding rare differences between data from different sources but for the same individual.

In general, my view is as long as read placement is perfect, even naive methods work sufficiently well. Complex methods only give you theoretical comfort in that case. One of hard parts is all kinds of alignment artifacts. SNVMix and another paper published in Bioinformatics last year try to resolve this by machine learning using multiple statistics that may imply alignment errors. I, however, always prefer to tackle a problem directly, if we can, rather than via machine learning. To me, the simplest yet most effective strategy is to use two distinct alignment algorithms, such as bwa and bwa-sw, which have distinct error modes. You only consider mutations shared between the two alignments. False calls will be vastly reduced. Using decoy sequence also helps around centromeres.

Another complication is structural variations, in which I am less experienced. In some sense, false mutations caused by structural variations are still indication of something different between normal and tumor.

In all, I think you do not need to worry about which software to use for detecting somatic mutations &#8211; anything reasonable is fine. You should pay more attention to **mismapping and structural variations**. My sanger colleagues from the Cancer Group would agree with me. They have published many high-profile papers.  
[http://www.biostars.org/p/19104/]  
5. What methods do you use for In/Del/SNP calling http://www.biostars.org/p/613/