---
title: Rip-Seq
author: 悟道
layout: post
permalink: /?p=2940
categories:
  - seq
tags:
  - makefile
---
<table>
  <tr cellpadding=0><td>
    热度:
  </td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td><td cellpadding=0><img src='http://210.75.224.29/wordpress/wp-content/plugins/statpresscn/images/sun_dark.gif' width=10 height=10 border=0 /></td></tr>
</table>

<pre class="brush: bash; title: Makefile for Rip-seq; version 1; notranslate" title="Makefile for Rip-seq; version 1">include ~/home/server-project/makefile.tm

rawn=m6a_P16 m6a_P22 m6a_P23
raws=$(addsuffix .fastq,  $(rawn))
fastqc=$(addsuffix .fastqc, $(rawn))
trans=/home/chentong/home/project/project5/ucsc/mm9.gene.refgene.gtf
transIndex=/home/chentong/home/server-project/trans_index/mm9/
mm9=~/home/server-project/bowtie2_index/mm9/mm9.fa

$(fastqc):
	fastqc $(basename $@).fq &gt;$(basename $@).log 2&gt;&1

line=m6a_P16
current_sample=P16_P22_P23

$(line).bwa.genome:
	echo "Begin $@---`date`" &gt;&gt;$@.log
	echo "Begin mapping---`date`" &gt;&gt;$@.log
	bwa aln -t 6 $(mm9Index) $(line).fq &gt;$@.sai 2&gt;&gt;$@.log
	bwa samse $(mm9Index) $@.sai $(line).fq &gt; $@.sam 2&gt;&gt;$@.log
	/bin/rm -f $@.sai &
	echo "End mapping---`date`" &gt;&gt;$@.log
	echo "Begin filter unmapped reads and sort---`date`" &gt;&gt;$@.log
	head -n 1000 $@.sam | grep '^@' &gt;$@.uniq.sam
	grep 'XT:A:U' $@.sam &gt;&gt;$@.uniq.sam
	samtools view -F 4 -ubS $@.uniq.sam \
		| samtools sort - $@.uniq.map.sort 2&gt;&gt;$@.log
	/bin/rm -f $@.uniq.sam &
	###echo "End filter unmapped reads and sort---`date`" &gt;&gt;$@.log
	###echo "Begin selecting high quality reads---`date`" &gt;&gt;$@.log
	###samtools view -F4 -q 1 -ubS $@.sam \
	###	| samtools sort - $@.map.q1.sort 2 &gt;&gt;$@.log
	###echo "End selecting high quality reads---`date`" &gt;&gt;$@.log
	###echo "Begin removing PCR artifacast---`date`" &gt;&gt;$@.log
	###samtools rmdup -s $@.map.q1.sort.bam \
	###	$@.map.q1.sort.rmdup.bam 2&gt;&gt;$@.log
	#samtools rmdup -s $@.uniq.map.sort.bam \
	#	$@.uniq.map.sort.rmdup.bam 2&gt;&gt;$@.log
	#echo "End removing PCR artifact---`date`" &gt;&gt;$@.log
	###echo "Begin indexing---`date`" &gt;&gt;$@.log
	###samtools index $@.map.q1.sort.rmdup.bam 2&gt;&gt;$@.log
	###echo "End indexing---`date`" &gt;&gt;$@.log
	echo "Total reads:`samtools view -cS $@.sam`" &gt;&gt;$@.sta
	/bin/rm -f $@.sam &
	#echo "Total redundant reads:`samtools view -c $@.map.q1.sort.bam`" &gt;&gt;$@.sta
	#bin/rm -f $@.map.q1.sort.bam &
	echo "Total unique redundant reads:`samtools view -c $@.uniq.map.sort.bam`" &gt;&gt;$@.sta
	#/bin/rm -f $@.uniq.map.sort.bam &
	#echo "Final reads:`samtools view -c $@.map.q1.sort.rmdup.bam`" &gt;&gt;$@.sta
	#echo "Final reads:`samtools view -c $@.uniq.map.sort.rmdup.bam`" &gt;&gt;$@.sta
	#echo "End $@---`date`" &gt;&gt;$@.log
	#echo "bamTobed `date`---Begins" &gt;&gt;$(basename $@).log
	#bamToBed -i $@.map.q1.sort.rmdup.bam | sort -k1,1 &gt;$@.map.q1.sort.rmdup.bed 2&gt;&gt;$@.log
	#slopBed -i $@.map.q1.sort.rmdup.bed -g ~/home/server-project/ucsc/hg19.chromInfo -r 268 -l 0 -s &gt;$@.map.q1.sort.rmdup.ext300.bed
	#echo "bamTobed `date`---Ends" &gt;&gt;$(basename $@).log
	touch $@
	pythonmail2.py $@ $@

$(line).bowtie.genome:
	echo "Begin mapping $@---`date`" &gt;&gt;$@.log
	bowtie -n2 -l28 -e70 --chunkmbs 300 -a -m1 -S --best --strata -p6 $(bowtiemm9) $(line).fq $@.sam&gt;$@.log 2&gt;&1 
	echo "End mapping $@---`date`" &gt;&gt;$@.log
	samtools view -F 4 -ubS $@.sam | samtools sort - $@.uniq.map.sort
	echo "Total reads:`samtools view -cS $@.sam`" &gt;&gt;$@.sta
	echo "Total unique redundant reads:`samtools view -c $@.uniq.map.sort.bam`" &gt;&gt;$@.sta
	/bin/rm -f $@.sam &
	pythonmail2.py $@ $@
#-----------------only for single strand--------------------
type=stdGTF
reads_clipping:
	extractFastqByName.py $(line).fq &lt;(samtools view $(line)/$(line).tophat_$(type)/unmapped.bam | cut -f 1) &gt;$(line).$(type).unmapped.fq
	deleteNinFASTQ.py &lt;(clipAdaptor.py $(line).$(type).unmapped.fq 'ACTGTAGGCACCATCAATC') \
		| fastq_quality_trimmer -o $(line).$(type).unmapped.clipped.delN.trimq20.fq -t 20 -l 30 -v -Q 33
	cat $(line).$(type).unmapped.clipped.delN.trimq20.fq &lt;(extractFastqByName.py $(line).fq &lt;(samtools view $(line)/$(line).tophat_$(type)/accepted_hits.bam | cut -f 1)) &gt;$(line).$(type).processed.fq
	pythonmail2.py $(line).$(type).processed.fq $(line).$(type).processed.fq 
#-----------------only for single strand--------------------
uniq=uniq
reads_len_distrib=yes
igv=igv

$(line).tophat_stdGTF:
	/bin/mkdir -p $(line)
	awk '{if(FNR%4==2) if($$0~/N/) sum +=1}END{print sum}' $(line).fq &gt; $(line)/$(line).readsCntN &
	tophat -o $(line)/$@ \
		--solexa-quals -p 8 \
		-G $(trans) --transcriptome-index $(transIndex) \
		--library-type fr-unstranded \
		-a 6 \
		$(bowtie2mm9) $(line).fq &gt;$@.log 2&gt;&1
ifeq ($(igv),igv)
	samtools index $(line)/$@/accepted_hits.bam 
	ln -sf `pwd`/$(line)/$@/accepted_hits.bam `pwd`/$(line)/$@/$(line).rip.bam 
	ln -sf `pwd`/$(line)/$@/accepted_hits.bam.bai `pwd`/$(line)/$@/$(line).rip.bam.bai
	igvtools count --includeDuplicates $(line)/$@/$(line).rip.bam $(line)/$@/$(line).rip.tdf,$(line)/$@/$(line).rip.wig mm9
	awk 'BEGIN{OFS="\t";FS="\t" }{if($$2!="") {print $$1,10^6*$$2/$($(line).map)} else print $$0}' $(line)/$@/$(line).rip.wig &gt;$(line)/$@/$(line).rip.wig.tmp && /bin/mv -f $(line)/$@/$(line).rip.wig.tmp $(line)/$@/$(line).rip.wig && igvtools toTDF $(line)/$@/$(line).rip.wig $(line)/$@/$(line).rip.tdf mm9 
endif
ifeq ($(stat),stat)
	samtools flagstat $(line)/$@/accepted_hits.bam &gt;$(line)/$@/$@.sta &
	echo "Total mapped reads:`samtools view $(line)/$@/accepted_hits.bam |  cut -f 1 | sort -u | wc -l`" &gt;$(line)/$@/mapped.sta &
	#---before wrongly ignore -w ---
	echo "Uniq mapped reads:`samtools view $(line)/$@/accepted_hits.bam | grep -w 'NH:i:1' | cut -f 1 | sort -u | wc -l`" &gt;&gt;$(line)/$@/mapped.sta &
endif
ifeq ($(cigar), cigar)
	samtools view $(line)/$@/accepted_hits.bam | cut -f 6 | sort | uniq -c | awk 'BEGIN{OFS="\t"}{print $$2,$$1}' &gt;$(line)/$@/$@.cigar && awk 'BEGIN{OFS="\t";FS="\t"}{gsub(/[0-9]/, "", $$1);a[$$1] +=$$2}END{for (i in a) print i,a[i]}' $(line)/$@/$@.cigar | sort -k2,2nr &gt;$(line)/$@/$@.cigar2 &
endif
ifeq ($(uniq), uniq)
	(samtools view -H $(line)/$@/accepted_hits.bam; samtools view $(line)/$@/accepted_hits.bam | grep -w 'NH:i:1') | samtools view -bS - -o $(line)/$@/accepted_hits.uniq.bam
ifeq ($(reads_len_distrib), yes)
	samtools view $(line)/$@/accepted_hits.uniq.bam | awk '{print length($$10)}' &gt;$(line)/$@/$@.accepted_hits.uniq.bam.reads_len_distrib && histogram.sh $(line)/$@/$@.accepted_hits.uniq.bam.reads_len_distrib 'Length distribution of mapped reads' &
endif
	samtools view -h $(line)/$@/accepted_hits.uniq.bam | extendReadsinBam.py - 100 | samtools view -ubS - | samtools sort - $(line)/$@/accepted_hits.uniq_extend
ifeq ($(igv),igv)
	samtools index $(line)/$@/accepted_hits.uniq_extend.bam 
	ln -sf `pwd`/$(line)/$@/accepted_hits.uniq_extend.bam `pwd`/$(line)/$@/$(line).rip.uniq_extend.bam 
	ln -sf `pwd`/$(line)/$@/accepted_hits.uniq_extend.bam.bai `pwd`/$(line)/$@/$(line).rip.uniq_extend.bam.bai
	igvtools count --includeDuplicates $(line)/$@/$(line).rip.uniq_extend.bam $(line)/$@/$(line).rip.uniq_extend.tdf,$(line)/$@/$(line).rip.uniq_extend.wig mm9
	awk 'BEGIN{OFS="\t";FS="\t" }{if($$2!="") {print $$1,10^6*$$2/$($@.unique)} else print $$0}' $(line)/$@/$(line).rip.uniq_extend.wig &gt;$(line)/$@/$(line).rip.uniq_extend.wig.tmp && /bin/mv -f $(line)/$@/$(line).rip.uniq_extend.wig.tmp $(line)/$@/$(line).rip.uniq_extend.wig && igvtools toTDF $(line)/$@/$(line).rip.uniq_extend.wig $(line)/$@/$(line).rip.uniq_extend.tdf mm9 &
endif #--end igv in uniq
	transferLong2ShortNameSpecialColumn.py &lt;(bamToBed -split -i $(line)/$@/accepted_hits.uniq_extend.bam) 4 0 0 | sort -k1,1 -k2,2n &gt; $(line)/$@/accepted_hits.uniq_extend.bed

endif #end ifeq (uniq)
ifeq ($(reads_len_distrib), yes)
	samtools view $(line)/$@/accepted_hits.bam | awk '{print length($$10)}' &gt;$(line)/$@/$(line).accepted_hits.bam.reads_len_distrib && histogram.sh $(line)/$@/$@.accepted_hits.bam.reads_len_distrib 'Length distribution of mapped reads' &
endif
	pythonmail2.py $@ $@

m6a_P16.stdGTF.processed.tophat_stdGTF.map=66997310
m6a_P16.stdGTF.processed.tophat_stdGTF.unique=63325505
m6a_P22.stdGTF.processed.tophat_stdGTF.map=115904994
m6a_P22.stdGTF.processed.tophat_stdGTF.unique=107319297
m6a_P23.stdGTF.processed.tophat_stdGTF.map=51986033
m6a_P23.stdGTF.processed.tophat_stdGTF.unique=48933476

mm9_gene_model=~/home/project/project5/ucsc/mm9.gene.refgene.bed
$(line).tophat_stdGTF.readsDistrib:
	read_distribution.py -i $(line)/$(basename $@)/accepted_hits.bam -r $(mm9_gene_model) &gt;$(line)/$(basename $@)/$@
	
#---for intersectBed,  two files must all be sorted by -k1,1 -k2,2n -----
mm9_gene_normal_model=~/home/project/project5/ucsc/mm9.gene.refgene.normal.bed
$(line).tophat_stdGTF.readsDistrib_self:
	#intersectBed -a $(line)/$(basename $@)/accepted_hits.uniq.bam -b $(mm9_gene_normal_model) -bed -wb -f 0.5 -split &gt;$(line)/$(basename $@)/$(basename $@).intersect
	#sort -k16 $(line)/$(basename $@)/$(basename $@).intersect | groupBy -i - -g 16 -c 4 -o collapse &gt;$(line)/$(basename $@)/$(basename $@).groupBy 
	#intersectBed -a $(line)/$(basename $@)/accepted_hits.uniq_extend.bed -b $(mm9_gene_normal_model) -wb -f 0.5 -sorted | awk 'BEGIN{OFS="\t";FS="\t"}{print $$10,$$4}' | sort -k1,1 &gt; $(line)/$(basename $@)/$(basename $@).intersect 
	#groupBy -i $(line)/$(basename $@)/$(basename $@).intersect -g 10 -c 4 -o collapse |  awk 'BEGIN{OFS="\t";FS="\t"}{a=split($$2,b, ","); print $$1,a}' &gt;$(line)/$(basename $@)/$@
	intersectBed -a $(line)/$(basename $@)/accepted_hits.uniq_extend.bed -b $(mm9_gene_normal_model) -wb -f 0.5 -sorted | awk 'BEGIN{OFS="\t";FS="\t"}{print $$10,$$4}' | sort -k1,1 | groupBy -i - -g 1 -c 2 -o collapse | awk 'BEGIN{OFS="\t";FS="\t"}{a=split($$2,b, ","); print $$1,a}' &gt;$(line)/$(basename $@)/$@
	awk 'BEGIN{OFS="\t";FS="\t"}{a=split($$1,b,"."); if(a==1) ele="Genebody"; else ele=b[2]; if (sum[ele] != "") sum[ele]+=$$2; else sum[ele]=$$2;}END{for (ele in sum) print ele,sum[ele]}' $(line)/$(basename $@)/$@ &gt;$(line)/$(basename $@)/$@.sum


mm9_bingenome=~/home/project/project5/ucsc/mm9.chrom.sizes.bin100_step50

$(line).tophat_stdGTF.bingenome:
	#coverageBed -split -abam $(line)/$(basename $@)/accepted_hits.uniq.bam -b $(mm9_bingenome) | cut -f 4,5,7 | awk 'BEGIN{OFS="\t";FS="\t"}{print $$1,10^9*$$2/$$3/$($(basename $@).unique)}' &gt;$(line)
	#coverageBed -split -counts -abam $(line)/$(basename $@)/accepted_hits.uniq.bam -b $(mm9_bingenome) | &gt;$(line)/$(basename $@)/mm9.bin100_step50.cnt
	coverageBed -counts -a $(line)/$(basename $@)/accepted_hits.uniq_extend.bed -b $(mm9_bingenome) | awk 'BEGIN{OFS="\t";FS="\t"}{print $$4,$$5,$$3-$$2}'&gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.cnt

ctl_sample=
ctl_cnt=~/home/server-project/m6a/BFC2011155/$(ctl_sample)/topout/$(ctl_sample).tophat.corrected.bin100_step50.cnt

TP16.ctl.unique=63069269
TP22.ctl.unique=69996091
TP23.ctl.unique=71681742

#----variable-------------
test_label_consistency=

$(line).tophat_stdGTF.peakcalling:
ifeq ($(test_label_consistency), yes)
	diff &lt;(cut -f 1 $(line)/$(basename $@)/$(basename $@).bin100_step50.cnt) &lt;(cut -f 1 $(ctl_cnt))
endif
	paste $(line)/$(basename $@)/$(basename $@).bin100_step50.cnt $(ctl_cnt) | cut -f 1,2,5 | sed '1 i\Region\tm6A\tCtl' | grep -vP '\t0\t0$$' &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.rr
ifneq ($(ctl_sample), )
	fisher.test.py $(line)/$(basename $@)/$(basename $@).bin100_step50.rr $($(basename $@).unique) $($(ctl_sample).ctl.unique) &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p
	contingencyTableCntTest.sh -f $(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p -n TRUE -z TRUE -s "$($(basename $@).unique),$($(ctl_sample).ctl.unique)"
	fisher.test.py $(line)/$(basename $@)/$(basename $@).bin100_step50.rr $($(basename $@).unique) $($(ctl_sample).ctl.unique) right_tail &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.rr.right_tail.p
	contingencyTableCntTest.sh -f $(line)/$(basename $@)/$(basename $@).bin100_step50.rr.right_tail.p -n TRUE -z TRUE -s "$($(basename $@).unique),$($(ctl_sample).ctl.unique)"
	awk 'BEGIN{OFS="\t";FS="\t"}{if(FNR!=1 && $$6-0.00001&lt;=0 && $$4-0.00001&lt;=0 && $$5&gt;=2 && $$2&gt;=20) $$NF="TRUE"; print $$0;}' $(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p.fisher.result &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p.fisher.result.sig
	awk 'BEGIN{OFS="\t";FS="\t"}{if(FNR!=1 && $$6-0.00001&lt;=0 && $$4-0.00001&lt;=0 && $$5&gt;=2 && $$2&gt;=20) $$NF="TRUE"; print $$0;}' $(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p.fisher.result &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.rr.p.fisher.result.sig
endif

$(line).tophat_stdGTF.gene_median_cnt:
	awk 'BEGIN{OFS="\t";FS="\t"}{a=split($$1,b,"_"); chr=b[2]; if(a==3) start=b[3]; else if(a==4) start=b[4]; print chr,start,start+$$3,$$1,$$2}' $(line)/$(basename $@)/$(basename $@).bin100_step50.cnt &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.cnt.bed
	intersectBed -a ~/home/project/project5/ucsc/mm9.gene.refgene.normal.gene.bed -b $(line)/$(basename $@)/$(basename $@).bin100_step50.cnt.bed -wa -wb | groupBy -i - -g 4 -c 11 -o median &gt;$(line)/$(basename $@)/$(basename $@).bin100_step50.gene_median_cnt

































#-------------------------------------------------------------------------

rna_label=$(subst m6a_,T,$(line))
selfTransAssemb=/home/chentong/home/server-project/m6a/BFC2011155/$(rna_label)/cufout_corrected/transcripts.gtf
newTransIndex=/home/chentong/home/server-project/m6a/BFC2011155/$(rna_label)/cufout_corrected/transcripts_index
$(line).tophat_selfGTF:
	/bin/mkdir -p $(line)
	tophat -o $(line)/$@ \
		--solexa-quals -p 8 \
		-G $(selfTransAssemb) --transcriptome-index $(newTransIndex) \
		--library-type fr-unstranded \
		$(bowtie2mm9) $(line).fq &gt;$@.log 2&gt;&1
	samtools flagstat $(line)/$@/accepted_hits.bam &gt;$(line)/$@/$@.sta &
	echo "Total mapped reads:`samtools view $(line)/$@/accepted_hits.bam |  cut -f 1 | sort -u | wc -l`" &gt;$(line)/$@/mapped.sta &
	#--beofre wrongly left -w-----------------------
	echo "Uniq mapped reads:`samtools view $(line)/$@/accepted_hits.bam | grep -w 'NH:i:1' | cut -f 1 | sort -u | wc -l`" &gt;&gt;$(line)/$@/mapped.sta &
	samtools view $(line)/$@/accepted_hits.bam | cut -f 6 | sort | uniq -c | awk 'BEGIN{OFS="\t"}{print $$2,$$1}' &gt;$(line)/$@/$@.cigar && sed 's/[0-9]//g' $(line)/$@/$@.cigar | sort | uniq -c |awk 'BEGIN{OFS="\t"}{print $$2,$$1}' &gt;$(line)/$@/$@.cigar2 &
	pythonmail2.py $@ $@


$(line).cufflinks:
	cufflinks -p 8 --GTF-guide $(trans) --label $(line) \
		-o $(line)/cufout $(line)/topout/accepted_hits.bam \
		--library-type fr-secondstrand &gt;$(line).log 2&gt;&1
	pythonmail2.py $@ $@

cuffmerge:
	cuffmerge -g $(trans) -s $(bowtie2mm9).fa -p 8 assemblies.txt


cuffdiff:
	cuffdiff -o TP16-22-23 -b $(mm9) -p 8 -L TP16,TP22,TP23 -u \
		--library-type fr-secondstrand \
		merged_asm/merged.gtf \
		TP16/topout/accepted_hits.bam \
		TP22/topout/accepted_hits.bam \
		TP23/topout/accepted_hits.bam
		

all:
	/bin/rm -rf $(line)
	/bin/mkdir $(line)
	make $(line).tophat
	make $(line).cufflinks

</pre>