---
title: Pandas usage
author: ct
layout: post
description:
modified:
categories: [Python]
tags: [python]
---
    
### What is pandas

Pandas是python中用于处理矩阵样数据的功能强大的包，提供了R中的`dataframe`和`vector`的操作，使得我们在使用python时，也可以方便、简单、快捷、高效地进行矩阵数据处理。

具体介绍详见<http://pandas.pydata.org/>。


* A fast and efficient **DataFrame** object for data manipulation with integrated indexing;
* Tools for reading and writing data between in-memory data structures and different formats: CSV and text files, Microsoft Excel, SQL databases, and the fast **HDF5** format;
* Intelligent **data alignment** and integrated handling of missing data: gain automatic label-based alignment in computations and easily manipulate **messy data into an orderly form**;
* Flexible **reshaping** and **pivoting** of data sets;
* Intelligent label-based slicing, fancy indexing, and subsetting of large data sets;
* Columns can be inserted and deleted from data structures for size mutability;
* Aggregating or transforming data with a powerful group by engine allowing split-apply-combine operations on data sets;
* High performance **merging** and **joining** of data sets;
* Hierarchical axis indexing provides an intuitive way of working with high-dimensional data in a lower-dimensional data structure;
* Time series-functionality: date range generation and frequency conversion, moving window statistics, moving window linear regressions, date shifting and lagging. Even create domain-specific time offsets and join time series without losing data;
* Highly optimized for performance, with critical code paths written in Cython or C.
* Python with pandas is in use in a wide variety of academic and commercial domains, including Finance, Neuroscience, Economics, Statistics, Advertising, Web Analytics, and more.



```python
%matplotlib inline

#import plotly
#plotly.offline.init_notebook_mode()

import matplotlib
matplotlib.style.use('ggplot')
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
from glob import glob
```

### Pandas读取文件

#### 获取目标文件


```python
dir_1 = "pandas_data/"
glob(dir_1+'*')
```




```python
['pandas_data/ENCFF289HGQ.tsv',
 'pandas_data/gencode.v24.ENS2SYN',
 'pandas_data/ENCFF262OBL.tsv',
 'pandas_data/Gene_metadata_primary_wt_whole_cell.tsv',
 'pandas_data/ENCFF673KYR.tsv',
 'pandas_data/ENCFF060LPA.tsv']
```



#### 查看目标文件内容和格式
Ipython中可以通过在Linux命令前加`!`调用系统命令，更多使用见
<http://ipython.org/ipython-doc/3/interactive/reference.html#system-shell-access>.


```python
!head -n 4 pandas_data/gencode.v24.ENS2SYN
```

```python
gene_id	gene_symbol
ENSG00000223972.5	DDX11L1
ENSG00000227232.5	WASH7P
ENSG00000278267.1	MIR6859-1
```



```python
!head -n 4 pandas_data/ENCFF556YSD.tsv
```

```python
transcript_id	gene_id	length	effective_length	expected_count	TPM	FPKM
ENST00000373020.4	ENSG00000000003.10	2206	1925.57	997.23	1.71	7.21
ENST00000494424.1	ENSG00000000003.10	820	539.58	24.77	0.15	0.64
ENST00000496771.1	ENSG00000000003.10	1025	744.57	0.00	0.00	0.00
```


#### 读取两列文件


```python
ens2syn_file = "pandas_data/gencode.v24.ENS2SYN"
```


```python
# pandas中的计数都是从0开始的
# header=0: 指定第一行包含列的名字
# index_col=0: 指定第一列为行的名字
ens2syn = pd.read_table(ens2syn_file, header=0, index_col=0)
```

#### 数据表的索引

* 数值索引和布尔值索引是按行选取
* 字符串索引是按列选取
* 行和列是等效的，应用于行的选取函数也可应用于列，反之亦然

##### 按行选取数据


```python
ens2syn[:3]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000223972.5</th>
      <td>DDX11L1</td>
    </tr>
    <tr>
      <th>ENSG00000227232.5</th>
      <td>WASH7P</td>
    </tr>
    <tr>
      <th>ENSG00000278267.1</th>
      <td>MIR6859-1</td>
    </tr>
  </tbody>
</table>
</div>



##### 取出索引中包含特定值的行


```python
ens2syn[ens2syn.index=="ENSG00000227232.5"]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000227232.5</th>
      <td>WASH7P</td>
    </tr>
  </tbody>
</table>
</div>



##### 取出某列包含特定值列表的行


```python
ens2syn[ens2syn['gene_symbol'].isin(['DDX11L1','MIR6859-1'])]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000223972.5</th>
      <td>DDX11L1</td>
    </tr>
    <tr>
      <th>ENSG00000278267.1</th>
      <td>MIR6859-1</td>
    </tr>
  </tbody>
</table>
</div>



##### 使用正则表达式选取符合要求的行


```python
# head: 只展示部分数据
ens2syn[ens2syn.index.str.contains(r'ENSG0000022')].head()
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000223972.5</th>
      <td>DDX11L1</td>
    </tr>
    <tr>
      <th>ENSG00000227232.5</th>
      <td>WASH7P</td>
    </tr>
    <tr>
      <th>ENSG00000222623.1</th>
      <td>RNU6-1100P</td>
    </tr>
    <tr>
      <th>ENSG00000228463.9</th>
      <td>AP006222.2</td>
    </tr>
    <tr>
      <th>ENSG00000224813.3</th>
      <td>SEPT14P13</td>
    </tr>
  </tbody>
</table>
</div>



#### 读取多列文件

`gzip`, `bzip`压缩的文件也可以直接读取，但是需要保证文件后缀的正确。
`read_table`默认参数可以自动检测文件的格式，根据文件的后缀 '.gz', '.bz2', '.zip', or 'xz'分别使用 gzip, bz2, zip or xz读取。


```python
tsvL = glob(dir_1+'ENC*.tsv')
tsvL
```




```python
['pandas_data/ENCFF289HGQ.tsv',
 'pandas_data/ENCFF262OBL.tsv',
 'pandas_data/ENCFF673KYR.tsv',
 'pandas_data/ENCFF060LPA.tsv']
```




```python
index = 0
tsvFile = tsvL[index]
expr = pd.read_table(tsvFile, header=0, index_col=0)
expr.head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>transcript_id(s)</th>
      <th>length</th>
      <th>effective_length</th>
      <th>expected_count</th>
      <th>TPM</th>
      <th>FPKM</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>ENST00000373020.8,ENST00000494424.1,ENST000004...</td>
      <td>2198.69</td>
      <td>1939.24</td>
      <td>2827.0</td>
      <td>1.03</td>
      <td>10.84</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>ENST00000373031.4,ENST00000485971.1</td>
      <td>940.50</td>
      <td>681.07</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>ENST00000371582.8,ENST00000371584.8,ENST000003...</td>
      <td>1079.84</td>
      <td>820.38</td>
      <td>1680.0</td>
      <td>1.45</td>
      <td>15.23</td>
    </tr>
  </tbody>
</table>
</div>



#### 选取多列数据
列的输出顺序与给定的列名字的顺序一致


```python
expr[['FPKM','TPM']].head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>FPKM</th>
      <th>TPM</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>10.84</td>
      <td>1.03</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>15.23</td>
      <td>1.45</td>
    </tr>
  </tbody>
</table>
</div>



#### 重命名列名字
从Dataframe中只选取一列时，数据框会被转换成Series，因此需要使用`pd.loc[:,[column_name]]`(虽然内部的方括号内只有一个值，但写法是必须的)索引。


```python
# 因为要把多个文件的同一类型表达值合并到一个文件，我们使用文件名作为列的名字
name = os.path.split(tsvFile)[-1][:-4]
print name
expr_tpm = expr.loc[:,['TPM']] # 取出所有的行和名字为TPM的列
expr_tpm.columns=[name]  
expr_tpm[:3]
```

```python
ENCFF289HGQ
```





<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
    </tr>
  </tbody>
</table>
</div>



#### 合并矩阵

##### 定义函数简化文件读取


```python
# 为了读取多个文件，定义一个函数简化操作
def readExpr_1(tsvFileL, typeL=['TPM','FPKM']):
    '''
    tsvFileL: lists of files waiting for reading
    resultD: a dictionary to save data matrix
            {'TPM':[mat1, mat2,...]
             'FPKM':[mat1, mat2, ...]}
    typeL; list of names for columns to be extracted
    '''
    resultD = {}
    for _type in typeL: resultD[_type] = []
    
    for tsvFile in tsvFileL:
        expr = pd.read_table(tsvFile, header=0, index_col=0)
        name = os.path.split(tsvFile)[-1][:-4]  #this options is very arbitary
        for _type in typeL: # add _ to type to avoid override Python inner function `type` 
            expr_type = expr.loc[:,[_type]]
            expr_type.columns = [name]
            resultD[_type].append(expr_type)
    return resultD
#-----------------------------------------------------
```


```python
exprD = readExpr_1(tsvL)
TPM_mat = exprD['TPM']
FPKM_mat = exprD['FPKM']
```

##### 使用pd.merge合并矩阵示例

先从刚才读取的矩阵中选出2个测试下pandas中的矩阵合并方法和效果


```python
# 选取第一个矩阵
_idL = ['ENSG00000000003.14', 'ENSG00000000005.5','ENSG00000000419.12','ENSG00000000457.13']
mat1 = TPM_mat[0]
mat1 = mat1[mat1.index.isin(_idL)]
mat1
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
    </tr>
  </tbody>
</table>
</div>




```python
# 选取第二个矩阵
_idL = ['ENSG00000001561.6','ENSG00000000003.14', 'ENSG00000000419.12','ENSG00000001036.13']
mat2 = TPM_mat[1]
mat2 = mat2[mat2.index.isin(_idL)]
mat2
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>



基于索引(index)的合并
```python
* outer: 合并所有的索引，缺失值填充NA
* inner：保留共有的索引
* left：使用第一个矩阵的索引
* right：使用第二个矩阵的索引
```


```python
pd.merge(mat1, mat2, left_index=True, right_index=True, how="outer")
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>NaN</td>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>NaN</td>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>




```python
pd.merge(mat1, mat2, left_index=True, right_index=True, how="inner")
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
  </tbody>
</table>
</div>




```python
pd.merge(mat1, mat2, left_index=True, right_index=True, how="left")
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>



##### 使用pd.concat合并矩阵示例 
对于较多的数据表合并操作时，`concat`比`merge`要简单快速很多。


```python
pd.concat([mat1, mat2], axis=1)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>NaN</td>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>NaN</td>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>




```python
pd.concat([mat1, mat2], axis=1, join="inner")
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
  </tbody>
</table>
</div>



##### 使用pd.join合并矩阵示例


```python
mat3 = mat1.join(mat2, how="outer")
mat3
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>NaN</td>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>NaN</td>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>



替换NA值为0


```python
mat3 = mat3.fillna(0)
mat3
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000005.5</th>
      <td>0.00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>0.00</td>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>0.00</td>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>



去除所有值都为0的行


```python
#Both works well here
#mat3[(mat3>0).any(axis=1)]
mat3.loc[(mat3>0).any(axis=1)]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>ENSG00000001036.13</th>
      <td>0.00</td>
      <td>10.34</td>
    </tr>
    <tr>
      <th>ENSG00000001561.6</th>
      <td>0.00</td>
      <td>2.47</td>
    </tr>
  </tbody>
</table>
</div>



##### 测试三种方法使用的内存和速度比较

速度：`concat`>`join`>>`merge`

内存：相当


```python
%timeit test_merge = reduce(lambda left,right: pd.merge(left,right,left_index=True,right_index=True,how='outer'), TPM_mat)
```

```python
100 loops, best of 3: 3.36 ms per loop
```



```python
%timeit test_merge = pd.concat(TPM_mat, axis=1)
```

```python
1000 loops, best of 3: 1.21 ms per loop
```



```python
%timeit TPM_mat[0].join(TPM_mat[1:], how="outer")
```

```python
1000 loops, best of 3: 1.25 ms per loop
```



```python
%load_ext memory_profiler
```


```python
%memit test_merge = reduce(lambda left,right: pd.merge(left,right,left_index=True,right_index=True,how='outer'), TPM_mat)
```

```python
peak memory: 107.32 MiB, increment: 0.01 MiB
```



```python
%memit test_merge = pd.concat(TPM_mat, axis=1)
```

```python
peak memory: 107.30 MiB, increment: 0.10 MiB
```



```python
%memit TPM_mat[0].join(TPM_mat[1:], how="outer")
```

```python
peak memory: 107.32 MiB, increment: 0.00 MiB
```


##### 重写函数完成文件的读写和矩阵的合并


```python
# 读取多个文件，并且合并矩阵，定义一个函数简化操作
def concatExpr(tsvFileL, typeL=['TPM','FPKM']):
    '''
    tsvFileL: lists of files waiting for reading
    resultD: a dictionary to save data matrix
            {'TPM':[mat1, mat2,...]
             'FPKM':[mat1, mat2, ...]}
    typeL; list of names for columns to be extracted
    '''
    resultD = {}
    for _type in typeL: resultD[_type] = []
    
    for tsvFile in tsvFileL:
        expr = pd.read_table(tsvFile, header=0, index_col=0)
        name = os.path.split(tsvFile)[-1][:-4]  #this options is very arbitary
        for _type in typeL: # add _ to type to avoid override Python inner function `type` 
            expr_type = expr.loc[:,[_type]]
            expr_type.columns = [name]
            resultD[_type].append(expr_type)
    #-------------------------------------------
    mergeD = {}
    for _type in typeL:
        mergeM = pd.concat(resultD[_type], axis=1)
        mergeM = mergeM.fillna(0) # Substitute all NA with 0
        mergeM = mergeM.loc[(mergeM>0).any(axis=1)] # Delete aoo zero rows.
        mergeD[_type] = mergeM
    return mergeD
#-----------------------------------------------------
```


```python
typeL = ['TPM','FPKM']
exprD = concatExpr(tsvL, typeL)
TPM_mat = exprD['TPM']
FPKM_mat = exprD['FPKM']
```


```python
TPM_mat.head()
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
      <th>ENCFF673KYR</th>
      <th>ENCFF060LPA</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
      <td>2.42</td>
      <td>6.64</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
      <td>1.80</td>
      <td>9.91</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>2.48</td>
      <td>0.38</td>
      <td>0.86</td>
    </tr>
    <tr>
      <th>ENSG00000000460.16</th>
      <td>0.26</td>
      <td>5.36</td>
      <td>0.16</td>
      <td>1.51</td>
    </tr>
    <tr>
      <th>ENSG00000000938.12</th>
      <td>0.00</td>
      <td>0.05</td>
      <td>0.00</td>
      <td>0.01</td>
    </tr>
  </tbody>
</table>
</div>



#### 矩阵数据提取

只保留表达矩阵中存储的基因的`ID`和`Symbol`对照表


```python
# 回顾下数据格式
ens2syn.head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000223972.5</th>
      <td>DDX11L1</td>
    </tr>
    <tr>
      <th>ENSG00000227232.5</th>
      <td>WASH7P</td>
    </tr>
    <tr>
      <th>ENSG00000278267.1</th>
      <td>MIR6859-1</td>
    </tr>
  </tbody>
</table>
</div>




```python
ens2syn.shape
```




```python
(60725, 1)
```




```python
ens2syn = ens2syn[ens2syn.index.isin(TPM_mat.index)]
```


```python
ens2syn.shape
```




```python
(48, 1)
```




```python
ens2syn.head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000001460.17</th>
      <td>STPG1</td>
    </tr>
    <tr>
      <th>ENSG00000001461.16</th>
      <td>NIPAL3</td>
    </tr>
    <tr>
      <th>ENSG00000000938.12</th>
      <td>FGR</td>
    </tr>
  </tbody>
</table>
</div>



#### 读取META data文件


```python
meta = "pandas_data/meta.tsv"
metaM = pd.read_table(meta, header=0, index_col=0)
# 重名了列的名字
oriColnames = metaM.columns.values
nameD = dict([(i,i.replace(' ','_')) for i in oriColnames])
metaM.rename(columns=nameD, inplace=True)
metaM.head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>File format</th>
      <th>Output type</th>
      <th>Experiment accession</th>
      <th>Assay</th>
      <th>Biosample term id</th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
      <th>Biosample organism</th>
      <th>...</th>
      <th>md5sum</th>
      <th>File download URL</th>
      <th>Assembly</th>
      <th>Platform</th>
      <th>Controlled by</th>
      <th>File Status</th>
      <th>Audit WARNING</th>
      <th>Audit INTERNAL_ACTION</th>
      <th>Audit NOT_COMPLIANT</th>
      <th>Audit ERROR</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF120PLK</th>
      <td>tsv</td>
      <td>gene quantifications</td>
      <td>ENCSR198TKA</td>
      <td>RNA-seq</td>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
      <td>Homo sapiens</td>
      <td>...</td>
      <td>1e9a3db25f5361b2ca454d1df427f430</td>
      <td>https://www.encodeproject.org/files/ENCFF120PL...</td>
      <td>hg19</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>released</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENCFF805BVE</th>
      <td>tsv</td>
      <td>gene quantifications</td>
      <td>ENCSR198TKA</td>
      <td>RNA-seq</td>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
      <td>Homo sapiens</td>
      <td>...</td>
      <td>ee0e94d6795ed7c2ef69c61b1d29eb02</td>
      <td>https://www.encodeproject.org/files/ENCFF805BV...</td>
      <td>hg19</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>released</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENCFF850RHD</th>
      <td>tsv</td>
      <td>gene quantifications</td>
      <td>ENCSR198TKA</td>
      <td>RNA-seq</td>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
      <td>Homo sapiens</td>
      <td>...</td>
      <td>22f948135c0935516f19f6b995ccc30c</td>
      <td>https://www.encodeproject.org/files/ENCFF850RH...</td>
      <td>GRCh38</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>released</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 47 columns</p>
</div>



##### 只保留前面提到的4个样品的数据


```python
sampleL = TPM_mat.columns.values
metaM = metaM[metaM.index.isin(sampleL)]
# 同时索引行和列
metaM.ix[:4,:5]
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Biosample term id</th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF673KYR</th>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>CL:0002558</td>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>male, female</td>
    </tr>
  </tbody>
</table>
</div>



#####  提取目标列信息


```python
# 假如只提取`Biosample`开头的列
#meta_colL = ['Biosample term id', 'Biosample term name']

# Extract columns matching specific patterns
# Both works well, filter is more simple
#metaM.loc[:,metaM.columns.str.contains(r'^Biosample')]
metaM = metaM.filter(regex=("^Biosample"))
metaM
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Biosample term id</th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
      <th>Biosample organism</th>
      <th>Biosample treatments</th>
      <th>Biosample subcellular fraction term name</th>
      <th>Biosample phase</th>
      <th>Biosample synchronization stage</th>
      <th>Biosample Age</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF673KYR</th>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
      <td>Homo sapiens</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
      <td>Homo sapiens</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>52 year</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
      <td>Homo sapiens</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>23 year</td>
    </tr>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>CL:0002558</td>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>male, female</td>
      <td>Homo sapiens</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>




```python
metaM.fillna('')
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Biosample term id</th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
      <th>Biosample organism</th>
      <th>Biosample treatments</th>
      <th>Biosample subcellular fraction term name</th>
      <th>Biosample phase</th>
      <th>Biosample synchronization stage</th>
      <th>Biosample Age</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF673KYR</th>
      <td>CL:0000650</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
      <td>Homo sapiens</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
      <td>Homo sapiens</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>52 year</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>CL:1001568</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
      <td>Homo sapiens</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>23 year</td>
    </tr>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>CL:0002558</td>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>male, female</td>
      <td>Homo sapiens</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div>



### Pandas写入文件

#### 写入文本文件


```python
metaM.to_csv("pandas_data/meta2.tsv", sep="\t")
```


```python
ens2syn.to_csv("pandas_data/gencode.v24.ENS2SYN", sep="\t")
```


```python
TPM_mat.to_csv("pandas_data/TPM", sep=b'\t', float_format="%.2f")
```

#### 写入HDF5文件

HDF5 is a data model, library, and file format for storing and managing data. It supports an unlimited variety of datatypes, and is designed for flexible and efficient I/O and for high volume and complex data. HDF5 is portable and is extensible, allowing applications to evolve in their use of HDF5. The HDF5 Technology suite includes tools and applications for managing, manipulating, viewing, and analyzing data in the HDF5 format. <https://support.hdfgroup.org/HDF5/>

使用优势是把处理好的数据以二进制文件存取，既可以减少文件数目、压缩使用空间，又可以方便多次快速读取，并且可以在不同的程序语言如Python与R中共同使用。

HDF5文件的写入形式上类似于字典操作，其读取也是。


```python
# 写入模式打开一个HDF5文件,使用压缩格式以节省空间
store = pd.HDFStore("pandas_data/ENCODE.hdf5", "w", complib=str("zlib"), complevel=9)

# 写入表达矩阵
store["TPM"] = TPM_mat
store["FPKM"] = FPKM_mat

# 写入注释文件
store['ens2syn'] = ens2syn
store['meta'] = metaM

# 关闭HDF5句柄
store.close()
```

```python
/MPATHB/soft/anacond/lib/python2.7/site-packages/IPython/core/interactiveshell.py:3035: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->mixed,key->block0_values] [items->['Biosample term id', 'Biosample term name', 'Biosample type', 'Biosample life stage', 'Biosample sex', 'Biosample organism', 'Biosample Age']]

  exec(code_obj, self.user_global_ns, self.user_ns)
```


当数据中存在混合数据模式时，会出现上面的Warning，对于我们的数据只要把`metaM`中的`NaN`值替换掉就可以。


```python
# 写入模式打开一个HDF5文件,使用压缩格式已节省空间
store = pd.HDFStore("pandas_data/ENCODE.hdf5", "w", complib=str("zlib"), complevel=9)

# 写入表达矩阵
store["TPM"] = TPM_mat
store["FPKM"] = FPKM_mat

# 写入注释文件
store['ens2syn'] = ens2syn
store['meta'] = metaM.fillna('')

# 关闭HDF5句柄
store.close()
```

#### 读取HDF5文件


```python
store = pd.HDFStore("pandas_data/ENCODE.hdf5")
```


```python
# 列出HDF5文件的索引名字
store.keys()
```




```python
['/FPKM', '/TPM', '/ens2syn', '/meta']
```




```python
TPM_mat = store['TPM']
TPM_mat.head(3)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
      <th>ENCFF673KYR</th>
      <th>ENCFF060LPA</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000000003.14</th>
      <td>1.03</td>
      <td>17.13</td>
      <td>2.42</td>
      <td>6.64</td>
    </tr>
    <tr>
      <th>ENSG00000000419.12</th>
      <td>1.45</td>
      <td>18.86</td>
      <td>1.80</td>
      <td>9.91</td>
    </tr>
    <tr>
      <th>ENSG00000000457.13</th>
      <td>0.24</td>
      <td>2.48</td>
      <td>0.38</td>
      <td>0.86</td>
    </tr>
  </tbody>
</table>
</div>




```python
ens2syn = store['ens2syn']
meta = store['meta']
```


```python
store.close()
```

### PANDAS矩阵的小应用

利用上面的矩阵操作，选取这两个基因相关的信息并绘制表达谱


```python
targetL = ['KRIT1','AK2']
```

Gene_symbol转换为Gene_id


```python
ensID = ens2syn[ens2syn["gene_symbol"].isin(targetL)]
ensID
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_symbol</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000004455.16</th>
      <td>AK2</td>
    </tr>
    <tr>
      <th>ENSG00000001631.14</th>
      <td>KRIT1</td>
    </tr>
  </tbody>
</table>
</div>



提取目标基因的表达


```python
targetExpr = TPM_mat[TPM_mat.index.isin(ensID.index)]
targetExpr
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
      <th>ENCFF673KYR</th>
      <th>ENCFF060LPA</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENSG00000001631.14</th>
      <td>1.15</td>
      <td>13.36</td>
      <td>1.37</td>
      <td>6.21</td>
    </tr>
    <tr>
      <th>ENSG00000004455.16</th>
      <td>2.31</td>
      <td>37.62</td>
      <td>8.95</td>
      <td>15.57</td>
    </tr>
  </tbody>
</table>
</div>



重命名矩阵的索引


```python
ensID_dict = ensID.to_dict()
ensID_dict
```




```python
{'gene_symbol': {'ENSG00000001631.14': 'KRIT1', 'ENSG00000004455.16': 'AK2'}}
```




```python
targetExpr = targetExpr.rename(index=ensID_dict['gene_symbol'])
targetExpr
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENCFF289HGQ</th>
      <th>ENCFF262OBL</th>
      <th>ENCFF673KYR</th>
      <th>ENCFF060LPA</th>
    </tr>
    <tr>
      <th>gene_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>KRIT1</th>
      <td>1.15</td>
      <td>13.36</td>
      <td>1.37</td>
      <td>6.21</td>
    </tr>
    <tr>
      <th>AK2</th>
      <td>2.31</td>
      <td>37.62</td>
      <td>8.95</td>
      <td>15.57</td>
    </tr>
  </tbody>
</table>
</div>



转置矩阵以增加META信息


```python
targetExpr_t = targetExpr.T
targetExpr_t
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>gene_id</th>
      <th>KRIT1</th>
      <th>AK2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>1.15</td>
      <td>2.31</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>13.36</td>
      <td>37.62</td>
    </tr>
    <tr>
      <th>ENCFF673KYR</th>
      <td>1.37</td>
      <td>8.95</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>6.21</td>
      <td>15.57</td>
    </tr>
  </tbody>
</table>
</div>



从meta矩阵中提取4列信息


```python
meta_type = ["Biosample term name","Biosample type", "Biosample life stage", "Biosample sex"]
```


```python
meta = meta[meta_type]
meta
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF673KYR</th>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>unknown, fetal</td>
      <td>unknown, female</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>male, female</td>
    </tr>
  </tbody>
</table>
</div>



修改下矩阵信息，去除`unknow,`字符串（只是为了展示方便）


```python
meta.loc['ENCFF673KYR',"Biosample life stage"] = "fetal"
# Much faster
meta = meta.set_value('ENCFF673KYR','Biosample sex','female')
meta = meta.set_value('ENCFF289HGQ','Biosample sex','female')
meta
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
    </tr>
    <tr>
      <th>File accession</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF673KYR</th>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>fetal</td>
      <td>female</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>female</td>
    </tr>
  </tbody>
</table>
</div>




```python
target_expr_meta = targetExpr_t.join(meta, how="left")
target_expr_meta
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>gene_id</th>
      <th>KRIT1</th>
      <th>AK2</th>
      <th>Biosample term name</th>
      <th>Biosample type</th>
      <th>Biosample life stage</th>
      <th>Biosample sex</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>1.15</td>
      <td>2.31</td>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>female</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>13.36</td>
      <td>37.62</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF673KYR</th>
      <td>1.37</td>
      <td>8.95</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>fetal</td>
      <td>female</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>6.21</td>
      <td>15.57</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
  </tbody>
</table>
</div>



重名了列名字(替换掉名字中的空格)


```python
oriColnames = target_expr_meta.columns.values
nameD = dict([(i,i.replace(' ','_')) for i in oriColnames])
target_expr_meta.rename(columns=nameD, inplace=True)
target_expr_meta
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>gene_id</th>
      <th>KRIT1</th>
      <th>AK2</th>
      <th>Biosample_term_name</th>
      <th>Biosample_type</th>
      <th>Biosample_life_stage</th>
      <th>Biosample_sex</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENCFF289HGQ</th>
      <td>1.15</td>
      <td>2.31</td>
      <td>fibroblast of villous mesenchyme</td>
      <td>primary cell</td>
      <td>newborn</td>
      <td>female</td>
    </tr>
    <tr>
      <th>ENCFF262OBL</th>
      <td>13.36</td>
      <td>37.62</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
    <tr>
      <th>ENCFF673KYR</th>
      <td>1.37</td>
      <td>8.95</td>
      <td>mesangial cell</td>
      <td>primary cell</td>
      <td>fetal</td>
      <td>female</td>
    </tr>
    <tr>
      <th>ENCFF060LPA</th>
      <td>6.21</td>
      <td>15.57</td>
      <td>pulmonary artery endothelial cell</td>
      <td>primary cell</td>
      <td>adult</td>
      <td>male</td>
    </tr>
  </tbody>
</table>
</div>



绘制散点图


```python
target_expr_meta.plot.scatter(x='KRIT1', y='AK2')
```




```python
<matplotlib.axes._subplots.AxesSubplot at 0x7fbcaefc0c10>
```



```python
/MPATHB/soft/anacond/lib/python2.7/site-packages/matplotlib/collections.py:590: FutureWarning:

elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison

```



![png]({{ site.img_url}}/pandas/output_113_2.png)


绘制箱线图


```python
a = target_expr_meta.boxplot(["KRIT1", "AK2"])
```

```python
/MPATHB/soft/anacond/lib/python2.7/site-packages/IPython/kernel/__main__.py:1: FutureWarning:


The default value for 'return_type' will change to 'axes' in a future release.
 To use the future behavior now, set return_type='axes'.
 To keep the previous behavior and silence this warning, set return_type='dict'.

```



![png]({{ site.img_url }}/pandas/output_115_1.png)


使用Plotly绘制交互图


```python
fig = {
    'data': [
        {
            'x': target_expr_meta.KRIT1,
            'y': target_expr_meta.AK2,
            'text': target_expr_meta.Biosample_term_name,
            'mode': 'markers+texts',
            'name': 'Legend name',
            'marker': {
                'symbol':"circle",
                "opacity": "0.9"
            }
        },
        {
            'x': [0,40],
            'y': [0,40],
            'text': "Diagonal",
            'mode': 'lines',
            'name': 'Diagonal line',
            'showlegend': False,
            'line': {
                'color': ('rgb(192,192,192)')
            }
        }
    ],
    'layout': {
        'xaxis': {'title':'KRIT1 expression value','range':[0,40]},
        'yaxis': {'title':'AK2 expression value','range':[0,40]},
        'height':500,
        'width':600,
        'showlegend': True,
         "legend": {
            "x": 0.6,
            "y": 1
        }
    }
}
plotly.offline.iplot(fig)
```


<div id="77e25d92-10f1-4f66-99d1-e2cb8400f276" style="height: 500px; width: 600px;" class="plotly-graph-div"></div><script type="text/javascript">require(["plotly"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL="https://plot.ly";Plotly.newPlot("77e25d92-10f1-4f66-99d1-e2cb8400f276", [{"name": "Legend name", "text": ["fibroblast of villous mesenchyme", "pulmonary artery endothelial cell", "mesangial cell", "pulmonary artery endothelial cell"], "y": [2.31, 37.62, 8.95, 15.57], "mode": "markers+texts", "marker": {"opacity": "0.9", "symbol": "circle"}, "x": [1.15, 13.36, 1.37, 6.21]}, {"showlegend": false, "name": "Diagonal line", "text": "Diagonal", "mode": "lines", "y": [0, 40], "x": [0, 40], "line": {"color": "rgb(192,192,192)"}}], {"showlegend": true, "yaxis": {"range": [0, 40], "title": "AK2 expression value"}, "height": 500, "width": 600, "xaxis": {"range": [0, 40], "title": "KRIT1 expression value"}, "legend": {"y": 1, "x": 0.6}}, {"linkText": "Export to plot.ly", "showLink": true})});</script>



```python
fig = {
    'data': [
        {
            'x': target_expr_meta[target_expr_meta['Biosample_sex']==atype].KRIT1,
            'y': target_expr_meta[target_expr_meta['Biosample_sex']==atype].AK2,
            'text': target_expr_meta[target_expr_meta['Biosample_sex']==atype].Biosample_term_name,
            'mode': 'markers+texts',
            'name': _type,
            'marker': {
                'symbol':"circle",
                "opacity": "0.9"
            }
        } for atype in ['female','male']
    ],
    'layout': {
        'xaxis': {'title':'KRIT1 expression value','range':[0,40]},
        'yaxis': {'title':'AK2 expression value','range':[0,40]},
        'height':500,
        'width':600,
        'showlegend': True,
         "legend": {
            "x": 0.6,
            "y": 1
        }
    }
}
plotly.offline.iplot(fig)
```


<div id="a43f18bb-c691-4e91-98e5-8f8d607c171e" style="height: 500px; width: 600px;" class="plotly-graph-div"></div><script type="text/javascript">require(["plotly"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL="https://plot.ly";Plotly.newPlot("a43f18bb-c691-4e91-98e5-8f8d607c171e", [{"name": "female", "text": ["fibroblast of villous mesenchyme", "mesangial cell"], "y": [2.31, 8.95], "mode": "markers+texts", "marker": {"opacity": "0.9", "symbol": "circle"}, "x": [1.15, 1.37]}, {"name": "female", "text": ["pulmonary artery endothelial cell", "pulmonary artery endothelial cell"], "y": [37.62, 15.57], "mode": "markers+texts", "marker": {"opacity": "0.9", "symbol": "circle"}, "x": [13.36, 6.21]}], {"showlegend": true, "yaxis": {"range": [0, 40], "title": "AK2 expression value"}, "height": 500, "width": 600, "xaxis": {"range": [0, 40], "title": "KRIT1 expression value"}, "legend": {"y": 1, "x": 0.6}}, {"linkText": "Export to plot.ly", "showLink": true})});</script>


### 使用R读取HDF5文件

```r
#R code for reading hdf5

> h5ls('test.hdf5')
  group          name       otype  dclass      dim
0         /          FPKM   H5I_GROUP                 
1     /FPKM         axis0 H5I_DATASET  STRING        3
2     /FPKM         axis1 H5I_DATASET  STRING    25135
3     /FPKM  block0_items H5I_DATASET  STRING        3
4     /FPKM block0_values H5I_DATASET   FLOAT  x 25135
5         /           TPM   H5I_GROUP                 
6      /TPM         axis0 H5I_DATASET  STRING        3
7      /TPM         axis1 H5I_DATASET  STRING    24025
8      /TPM  block0_items H5I_DATASET  STRING        3
9      /TPM block0_values H5I_DATASET   FLOAT  x 24025
10        /       ens2syn   H5I_GROUP                 
11 /ens2syn         axis0 H5I_DATASET  STRING        1
12 /ens2syn         axis1 H5I_DATASET  STRING    60725
13 /ens2syn  block0_items H5I_DATASET  STRING        1
14 /ens2syn block0_values H5I_DATASET    VLEN        1
15        /          meta   H5I_GROUP                 
16    /meta         axis0 H5I_DATASET  STRING       47
17    /meta         axis1 H5I_DATASET  STRING        3
18    /meta  block0_items H5I_DATASET  STRING       19
19    /meta block0_values H5I_DATASET   FLOAT      x 3
20    /meta  block1_items H5I_DATASET  STRING        2
21    /meta block1_values H5I_DATASET INTEGER      x 3
22    /meta  block2_items H5I_DATASET  STRING       26
23    /meta block2_values H5I_DATASET    VLEN        1
> TPM = h5read("test.hdf5", "/TPM")
> str(TPM)
List of 4
 $ axis0        : chr [1:3(1d)] "ENCFF673KYR" "ENCFF805ZGF" "ENCFF581ZEU"
 $ axis1        : chr [1:24025(1d)] "ENSG00000000003.14" "ENSG00000000005.5" "ENSG00000000419.12" "ENSG00000000457.13" ...
 $ block0_items : chr [1:3(1d)] "ENCFF673KYR" "ENCFF805ZGF" "ENCFF581ZEU"
 $ block0_values: num [1:3, 1:24025] 2.42 1.64 5.69 0 0 0.11 1.8 3.82 6.38 0.38 ...
> d <- TPM$block0_values
> rownames(d) <- TPM$axis1
Error in `rownames<-`(`*tmp*`, value = c("ENSG00000000003.14", "ENSG00000000005.5",  : 
  length of 'dimnames' [1] not equal to array extent
> d <- as.data.frame(TPM$block0_values)
> rownames(d) <- TPM$axis1
Error in `row.names<-.data.frame`(`*tmp*`, value = value) : 
  invalid 'row.names' length
> dims(d)
Error: could not find function "dims"
> dim(d)
[1]     3 24025
> d <- t(as.data.frame(TPM$block0_values))
> dim(d)
[1] 24025     3
> rownames(d) <- TPM$axis1
> colnames(d) <- TPM$axis0
> hed(d)
Error: could not find function "hed"
> head(d)
               ENCFF673KYR ENCFF805ZGF ENCFF581ZEU
ENSG00000000003.14        2.42        1.64        5.69
ENSG00000000005.5         0.00        0.00        0.11
ENSG00000000419.12        1.80        3.82        6.38
ENSG00000000457.13        0.38        0.57        1.17
ENSG00000000460.16        0.16        0.31        0.14
ENSG00000000938.12        0.00        0.03        0.00
```


### Pandas矩阵生成


```python
np.random.seed(1)
df = pd.DataFrame({"first": np.random.rand(100),
                   "second": np.random.rand(100),
                   "class": np.random.randint(0, 2, (100,))},
                   index=range(100))
df.head()
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>class</th>
      <th>first</th>
      <th>second</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0.417022</td>
      <td>0.326645</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>0.720324</td>
      <td>0.527058</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>0.000114</td>
      <td>0.885942</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>0.302333</td>
      <td>0.357270</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1</td>
      <td>0.146756</td>
      <td>0.908535</td>
    </tr>
  </tbody>
</table>
</div>

### Ipython notebook link

<https://github.com/Tong-Chen/notebook/blob/master/pandas.ipynb>

### Blog link

原文链接 <{{ site.url }}/2016/11/pandas/>
