---
title: OrthoMCL installation 
author: ct
layout: post
categories:
  - Bioinformatics
tags:
  - OrthoMCL
---

### OrthoMCL能做什么

Orthologs are homologs separated by **speciation** events. Paralogs are
homologs separated by **duplication** events. Detection of orthologs is
becoming much more important with the rapid progress in genome
sequencing.

OrthoMCL is a genome-scale algorithm for grouping orthologous protein
sequences. It provides not only groups shared by two or more
species/genomes, but also groups representing species-specific gene
expansion families. So it serves as an important utility for automated
eukaryotic genome annotation. OrthoMCL starts with reciprocal best
hits within each genome as potential in-paralog/recent paralog pairs
and reciprocal best hits across any two genomes as potential ortholog
pairs. Related proteins are interlinked in a similarity graph. Then
MCL (Markov Clustering algorithm, Van Dongen 2000; www.micans.org/mcl)
is invoked to split mega-clusters. This process is analogous to the
manual review in COG construction. MCL clustering is based on weights
between each pair of proteins, so to correct for differences in
evolutionary distance the weights are normalized before running MCL.

OrthoMCL is similar to the INPARANOID algorithm (Remm, Storm et al.
2001),  but is extended to cluster orthologs from multiple
species. OrthoMCL clusters are coherent with groups identified by EGO
(Lee,  Sultana et al. 2002),  and an analysis using EC number suggests
a high degree of reliability (Li,  Stoeckert et al. 2003).
	
In a recent assessment (Chen,  et al. 2007),  the performance of seven
widely used orthology detection algorithms,  representing three
kinds of strategies (phylogeny-based,  evolutionary distance-based
and BLAST-based),  are evaluated using the statistical
technique **Latent Class Analysis** (LCA). LCA is useful when there
are large data sets available but no gold standard. The results
show an overall trade-off between sensitivity and specificity
among these algorithms,  with INPARANOID and OrthoMCL as the two
best methods having both False Positive (FP) and False Negative
(FN) error rates lower than 20%.

### 安装和使用

* 统一配置环境变量，一劳永逸

	* 环境变量配置：在系统中新建目录 `~/bin`，将其完整路径加入到环境变量。 
		* 把`export PATH=${PATH}:~/bin` 加到 `~/.bashrc`中

	* PERL5LIB配置：在系统中新建目录 `~/perl5lib`，将其完整路径加入到环境变量。
		* 把`export PERL5LIB=${PERL5LIB}:~/perl5lib/`加到`~/.bashrc`中

	* 更新环境变量配置 `source ~/.bashrc`

* mcl安装

```
wget http://www.micans.org/mcl/src/mcl-latest.tar.gz
tar xvzf mcl-latest.tar.gz
cd mcl-latest
./configure --prefix=`pwd`/../mcl_bin
make
make install
ln -s `pwd`/../mcl_bin/bin/* ~/bin/
```

* orthoMCL安装

```
wget http://orthomcl.org/common/downloads/software/v2.0/orthomclSoftware-v2.0.9.tar.gz
tar xvzf orthomclSoftware-v2.0.9.tar.gz
cd orthomclSoftware-v2.0.9
ln -s `pwd`/bin/* ~/bin/
ln -s `pwd`/lib/perl/* ~/perl5lib
```

* 配置Mysql数据库

	* 安装mysql数据库
		* `yum install mysql mysql-server`

	* 设置mysql根用户的密码
		* `mysql -uroot`登录mysql数据库
		* 在mysql操作界面依次输入sql语句 
		
		  `SET PASSWORD=PASSWORD("passwd");`
		  
		  `FLUSH PRIVILEGES;`

	* 因为OrthoMCL运行时需要较大的存储空间，而我的根目录下空间不够，
	  因此需要更换数据库目录；**如果根目录下空间足够，则不需要这部分操
	  作。**
		  
		* 查看mysql服务 `service mysqld status`

		* 关掉mysql服务 `service mysqld stop`

		* 移动数据库目录到目标位置 

			* `mkdir ~/mysql; chown mysql:mysql ~/mysql`
			* `mv /var/lib/mysql/* ~/mysql/`
			* 在`/etc/my.cnf`文件中修改`datadir`为`~/mysql`
    * 修改`/etc/my.cnf`配置文件

      ~~~~~~
      [mysqld]
      datadir=~/mysql
      #[OPTIMIZATION]
      ##Set this value to 50% of available RAM if your environment permits.
      myisam_sort_buffer_size=60G

      ##[OPTIMIZATION]
      ##This value should be at least 50% of free hard drive space. Use
      #caution if setting it to 100% of free space however. Your hard disk
      #may fill up!
      myisam_max_sort_file_size=200G

      ##[OPTIMIZATION]
      ##Our default of 2G is probably fine for this value. Change this value
      #only if you are using a machine with little resources available. 
      read_buffer_size=2G
      ~~~~~~

	* 启动mysql服务 `service mysqld start`

		* 若启动失败，查看log文件 `/var/log/mysqld.log`中的错误信息。
	    * `/usr/libexec/mysqld: Can't change dir to [Error code 13]`
			* 确保datadir的所有上层目录有`x`属性
			* 若依然启动不了，在终端运行`setenforce 0` 关闭SELINUX

	* 新建用户和数据库
		* 新建名字为`orthomcl`的数据库 `CREATE DATABASE orthomcl;`
		* 新建用户`orthomcl`，密码为`152108`, 该用户对数据库`orthomcl`有完全操作
		  权限
		  
		  GRANT SELECT,INSERT,UPDATE,DELETE,CREATE VIEW,
		  CREATE,INDEX,DROP on orthomcl.* TO 'orthomcl'@'localhost' 
		  IDENTIFIED BY '152108';
		  
		  FLUSH PRIVILEGES;

	* centos7中使用mariadb取代了mysql, 但所有命令的执行相同 (忽略掉这一段)

		yum install mariadb mariadb-server

		systemctl start mariadb ==> 启动mariadb

		systemctl enable mariadb ==> 开机自启动

		mysql_secure_installation ==> 设置 root密码等相关

		mysql -uroot -pPASSWD ==> 测试登录！

* 配置OrthoMCL工作文件

	* 建一个目录 (~/orthmcl_work)，存储OrthoMCL配置文件
	* 拷贝orthomclSoftware-v2.0.9/doc/OrthoMCLEngine/Main/orthomcl.config.template
	  到~/orthmcl_work，重命名为orthomcl.config
	* 修改内容为：

```
# this config assumes a mysql database named 'orthomcl'. 
# Adjust according to your situation.
dbVendor=mysql 
#Databsename: orthmcl
dbConnectString=dbi:mysql:orthomcl
#Database username
dbLogin=orthomcl
#Database password
dbPassword=152108
# Change strings as you like
similarSequencesTable=SimilarSequences
orthologTable=Ortholog
inParalogTable=InParalog
coOrthologTable=CoOrtholog
#Standards
interTaxonMatchView=InterTaxonMatch
percentMatchCutoff=50
evalueExponentCutoff=-5
oracleIndexTblSpc=NONE
```
	* 生成数据表：
		* `orthomclInstallSchema orthomcl.config inst_schema.log species`

* 创建OrthoMCL输入文件

	* OrthoMCL的输入文件为fasta格式文件，其中fasta序列的名字格式为
	  `>taxoncode|unique_prot_id`。序列名称为空格或下划线分开的两列，
	  第一列为3到4个字母的物种代码，第二列为蛋白序列的唯一ID。
	* 通常一个基因选择一条代表性蛋白序列。
	* 这些文件使用统一后缀`.fasta`，并存储于同一文件夹`orthlMCL`下 
	  (这个文件夹下只能存储fasta格式序列，不然运行
	   `orthomclBlastParser`时会报错)。
	* 序列过滤，允许最短的蛋白长度为10，stop codons最大比例为20%,
	  默认得到`goodProteins.fasta`。
		* `orthomclFilterFasta orthlMCL 10 20`
	* ~~将得到的`goodProteins.fasta`与
	  [orthoMCL](http://orthomcl.org/common/downloads/)的数据合并，
	  得到`orthoMCL.fa`。~~
	* 通常我们需要准备研究物种及其多个近缘或者有代表性物种的蛋白质序列
	  ，因此可不与orthoMCL数据库中的蛋白质序列合并，直接用我们的
	  `goodProteins.fasta`作为`orthoMCL.fa`。
	

* 序列BLAST

```
makeblastdb -in orthoMCL.fa -dbtype prot -title orthomcl \
	-out orthomcl -logfile orthomcl.log`
blastp -db orthomcl -query goodProteins.fasta -seg yes \
	-out orthomcl.blastout -evalue 1e-5 -outfmt 7 -num_threads 70`
```

* 略却其它步骤，都整合到一个bash脚本中。

* 整合的分析脚本[`orthoMcl.sh`](https://github.com/Tong-Chen/NGS/blob/master/orthoMcl.sh)

```
Usage:

/MPATHB/self/NGS/orthoMcl.sh options

Function:

This script is used to perform orthoMcl analysis using MySql, MCL and
orthomcl.

Before running this script, one must have one mysql database and a
mysql user which can perform operation on this database.


OPTIONS:
	-d	Mysql database name (using user_name as prefix to avoid
		duplication) [Necessary]
	-u	Mysql database username [Necessary]
	-p	Mysql database password [Necessary]
	-s	Target species of this analysis 
		(Any representing string is OK, the shorter the better)
		[Necessary]
	-D	A directory containing FASTA files for all proteins.
		[Necessary]
	-S	Sequences downloaded from orthMCL website.
		[Optional, not used anymore]
	-t	Number of threads for blast. [Default 50]
```

